// @coveo/headless/dist/esm/controllers/quickview/case-assist-headless-quickview.js
import { logQuickviewDocumentSuggestionClick } from '../../features/case-assist/case-assist-analytics-actions.js';
import { documentSuggestionReducer as documentSuggestion } from '../../features/document-suggestion/document-suggestion-slice.js';
import { preparePreviewPagination } from '../../features/result-preview/result-preview-actions.js';
import { buildResultPreviewRequest } from '../../features/result-preview/result-preview-request-builder.js';
import { loadReducerError } from '../../utils/errors.js';
import { buildCoreQuickview, } from '../core/quickview/headless-core-quickview.js';
/**
 * Creates a `CaseAssistQuickview` controller instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `CaseAssistQuickview` properties.
 * @returns A `CaseAssistQuickview` controller instance.
 *
 * @group Controllers
 * @category CaseAssistQuickview
 */
export function buildCaseAssistQuickview(engine, props) {
    if (!loadSearchQuickviewReducers(engine)) {
        throw loadReducerError;
    }
    const { dispatch } = engine;
    const getState = () => engine.state;
    const getDocuments = () => getState().documentSuggestion.documents;
    const fetchResultContentCallback = () => {
        engine.dispatch(logQuickviewDocumentSuggestionClick(props.options.result.uniqueId));
    };
    const path = '/html';
    const core = buildCoreQuickview(engine, props, buildResultPreviewRequest, path, fetchResultContentCallback);
    dispatch(preparePreviewPagination({
        results: getDocuments(),
    }));
    return {
        ...core,
        get state() {
            return {
                ...core.state,
                currentDocument: getDocuments().findIndex((r) => r.uniqueId === core.state.currentResultUniqueId) + 1,
                totalDocuments: getDocuments().length,
            };
        },
    };
}
function loadSearchQuickviewReducers(engine) {
    engine.addReducers({ documentSuggestion });
    return true;
}
