// @coveo/headless/dist/esm/controllers/core/facets/_common/facet-id-generator.js
export function generateFacetId(config, logger) {
    const { field, state } = config;
    if (!isFieldUsedAsFacetId(config)) {
        return field;
    }
    const prefix = `${field}_`;
    const id = calculateId(prefix, state);
    logWarningMessage(field, logger);
    return `${prefix}${id}`;
}
function isFieldUsedAsFacetId(config) {
    const { field, state } = config;
    const sets = extractFacetSets(state);
    return sets.some((set) => set && field in set);
}
function calculateId(prefix, state) {
    const sets = extractFacetSets(state);
    const keys = sets
        .map((set) => Object.keys(set || {}))
        .reduce((all, current) => all.concat(current), []);
    return findMaxId(keys, prefix) + 1;
}
function extractFacetSets(state) {
    const { facetSet, numericFacetSet, dateFacetSet, categoryFacetSet } = state;
    return [facetSet, numericFacetSet, dateFacetSet, categoryFacetSet];
}
function findMaxId(keys, prefix) {
    const defaultId = 0;
    const ids = keys.map((key) => {
        const stringId = key.split(prefix)[1];
        const id = parseInt(stringId, 10);
        return Number.isNaN(id) ? defaultId : id;
    });
    const lastNumber = ids.sort().pop();
    return lastNumber ?? defaultId;
}
function logWarningMessage(field, logger) {
    const message = `A facet with field "${field}" already exists.
  To avoid unexpected behaviour, configure the #id option on the facet controller.`;
    logger.warn(message);
}
