// @coveo/headless/dist/esm/features/search-parameters/search-parameter-utils.js
import { isNullOrUndefined } from '@coveo/bueno';
import { isEmptyString } from '../../utils/utils.js';
import { buildDateRanges, buildNumericRanges, cast, facetSearchParamRegex, isFacetObject, isRangeFacetObject, isValidBasicKey, keyHasObjectValue, } from './search-parameter-serializer.js';
export function toArray(value) {
    return Array.isArray(value) ? value : [value];
}
export function addFacetValuesToSearchParams(facetId, paramKey) {
    return (searchParams, valueArray) => {
        if (paramKey in searchParams) {
            const record = (searchParams[paramKey] ?? {});
            record[facetId] = [...(record[facetId] ?? []), ...valueArray];
            searchParams[paramKey] = record;
        }
        else {
            searchParams[paramKey] = { [facetId]: valueArray };
        }
    };
}
export function isValidSearchParam(key) {
    return facetSearchParamRegex.exec(key) !== null || isValidBasicKey(key);
}
export function isFacetPair(pair) {
    const [key, value] = pair;
    const isValidKey = keyHasObjectValue(key);
    const isValidValue = isFacetObject(value);
    return isValidKey && isValidValue;
}
export function isRangeFacetPair(pair) {
    const [key, value] = pair;
    const isValidKey = key === 'nf' || key === 'df';
    const isValidValue = isRangeFacetObject(value);
    return isValidKey && isValidValue;
}
/**
 * Extends the search parameters with the given key-value pair. If the key-value pair is not valid or if the value is undefined, the search parameters are not modified.
 * @param searchParams - The search parameters object to extend.
 * @param key - The key of the search parameter.
 * @param value - The value of the search parameter.
 */
export function extendSearchParameters(searchParams, key, value) {
    if (isNullOrUndefined(value)) {
        return;
    }
    if (isValidBasicKey(key) &&
        typeof value === 'string' &&
        !isEmptyString(value)) {
        const [k, v] = cast([key, value], false);
        searchParams[k] = v;
        return;
    }
    const result = facetSearchParamRegex.exec(key);
    if (result) {
        const paramKey = result[1];
        const facetId = result[2];
        if (!keyHasObjectValue(paramKey)) {
            return;
        }
        const add = addFacetValuesToSearchParams(facetId, paramKey);
        const range = paramKey === 'nf'
            ? buildNumericRanges(toArray(value), 'selected')
            : paramKey === 'df'
                ? buildDateRanges(toArray(value), 'selected')
                : toArray(value);
        add(searchParams, range);
    }
}
