// @coveo/headless/dist/esm/controllers/case-field/headless-case-field.js
import { Schema } from '@coveo/bueno';
import { configuration } from '../../app/common-reducers.js';
import { logAutoSelectCaseField, logClassificationClick, logUpdateCaseField, } from '../../features/case-assist/case-assist-analytics-actions.js';
import { caseAssistConfigurationReducer as caseAssistConfiguration } from '../../features/case-assist-configuration/case-assist-configuration-slice.js';
import { fetchCaseClassifications, registerCaseField, updateCaseField, } from '../../features/case-field/case-field-actions.js';
import { caseFieldReducer as caseField } from '../../features/case-field/case-field-slice.js';
import { caseInputReducer as caseInput } from '../../features/case-input/case-input-slice.js';
import { fetchDocumentSuggestions } from '../../features/document-suggestion/document-suggestion-actions.js';
import { documentSuggestionReducer as documentSuggestion } from '../../features/document-suggestion/document-suggestion-slice.js';
import { loadReducerError } from '../../utils/errors.js';
import { requiredNonEmptyString, validateOptions, } from '../../utils/validate-payload.js';
import { buildController, } from '../controller/headless-controller.js';
const optionsSchema = new Schema({
    field: requiredNonEmptyString,
});
/**
 * Creates a `CaseField` controller instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `CaseField` controller properties.
 * @returns A `CaseField` controller instance.
 *
 * @group Controllers
 * @category CaseField
 */
export function buildCaseField(engine, props = {}) {
    if (!loadCaseFieldReducers(engine)) {
        throw loadReducerError;
    }
    const controller = buildController(engine);
    const { dispatch } = engine;
    const options = validateOptions(engine, optionsSchema, props.options, 'buildCaseField');
    const getState = () => {
        return engine.state;
    };
    const isRegistered = getState().caseField?.fields?.[options.field];
    if (!isRegistered) {
        dispatch(registerCaseField({
            fieldName: options.field,
            fieldValue: '',
        }));
    }
    return {
        ...controller,
        get state() {
            const loading = getState().caseField?.status?.loading ?? false;
            const error = getState().caseField?.status?.error ?? null;
            const field = getState().caseField?.fields?.[options.field];
            const value = field?.value ?? '';
            const suggestions = field?.suggestions ?? [];
            return {
                loading,
                error,
                value,
                suggestions,
            };
        },
        update(value, updatesToFetch, autoSelection) {
            const suggestionId = getState().caseField?.fields?.[options.field]?.suggestions?.find((s) => s.value === value)?.id;
            if (suggestionId) {
                autoSelection
                    ? dispatch(logAutoSelectCaseField(suggestionId))
                    : dispatch(logClassificationClick(suggestionId));
            }
            dispatch(updateCaseField({
                fieldName: options.field,
                fieldValue: value,
            }));
            !autoSelection && dispatch(logUpdateCaseField(options.field));
            updatesToFetch?.caseClassifications &&
                dispatch(fetchCaseClassifications());
            updatesToFetch?.documentSuggestions &&
                dispatch(fetchDocumentSuggestions());
        },
    };
}
function loadCaseFieldReducers(engine) {
    engine.addReducers({
        configuration,
        caseAssistConfiguration,
        caseInput,
        caseField,
        documentSuggestion,
    });
    return true;
}
