// @coveo/headless/dist/esm/features/facets/range-facets/numeric-facet-set/numeric-facet-selectors.js
import { baseFacetResponseSelector } from '../../facet-set/facet-set-selectors.js';
function isNumericFacetResponse(state, response) {
    return !!response && response.facetId in state.numericFacetSet;
}
export const numericFacetResponseSelector = (state, facetId) => {
    const response = baseFacetResponseSelector(state, facetId);
    if (isNumericFacetResponse(state, response)) {
        return response;
    }
    return undefined;
};
export const numericFacetActiveValuesSelector = (state, facetId) => {
    const facetResponse = numericFacetResponseSelector(state, facetId) || {
        values: [],
    };
    return facetResponse.values.filter((value) => value.state !== 'idle');
};
export const numericFacetSelectedValuesSelector = (state, facetId) => {
    const facetResponse = numericFacetResponseSelector(state, facetId) || {
        values: [],
    };
    return facetResponse.values.filter((value) => value.state === 'selected');
};
export const numericFacetExcludedValuesSelector = (state, facetId) => {
    const facetResponse = numericFacetResponseSelector(state, facetId) || {
        values: [],
    };
    return facetResponse.values.filter((value) => value.state === 'excluded');
};
