// @coveo/headless/dist/esm/controllers/insight/did-you-mean/headless-insight-did-you-mean.js
import { didYouMeanClick } from '../../../features/did-you-mean/did-you-mean-analytics-actions.js';
import { logDidYouMeanClick } from '../../../features/did-you-mean/did-you-mean-insight-analytics-actions.js';
import { executeSearch } from '../../../features/insight-search/insight-search-actions.js';
import { buildCoreDidYouMean, } from '../../core/did-you-mean/headless-core-did-you-mean.js';
const defaultDidYouMeanOptions = {
    automaticallyCorrectQuery: true,
    queryCorrectionMode: 'legacy',
};
/**
 * The insight DidYouMean controller is responsible for handling query corrections.
 * When a query returns no result but finds a possible query correction, the controller either suggests the correction or
 * automatically triggers a new query with the suggested term.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `DidYouMean` properties.
 *
 * @group Controllers
 * @category DidYouMean
 */
export function buildDidYouMean(engine, props = {
    options: defaultDidYouMeanOptions,
}) {
    const options = {
        ...defaultDidYouMeanOptions,
        ...props.options,
    };
    const controller = buildCoreDidYouMean(engine, { options });
    const { dispatch } = engine;
    return {
        ...controller,
        get state() {
            return controller.state;
        },
        applyCorrection() {
            controller.applyCorrection();
            dispatch(executeSearch({ legacy: logDidYouMeanClick(), next: didYouMeanClick() }));
        },
    };
}
