// @coveo/headless/dist/esm/controllers/field-suggestions/facet/headless-field-suggestions.js
import { configuration } from '../../../app/common-reducers.js';
import { updateFacetOptions } from '../../../features/facet-options/facet-options-actions.js';
import { executeFacetSearch, executeFieldSuggest, } from '../../../features/facets/facet-search-set/generic/generic-facet-search-actions.js';
import { specificFacetSearchSetReducer as facetSearchSet } from '../../../features/facets/facet-search-set/specific/specific-facet-search-set-slice.js';
import { registerFacet } from '../../../features/facets/facet-set/facet-set-actions.js';
import { facetExclude, facetSelect, logFacetExclude, logFacetSelect, } from '../../../features/facets/facet-set/facet-set-analytics-actions.js';
import { defaultFacetOptions, facetSetReducer as facetSet, } from '../../../features/facets/facet-set/facet-set-slice.js';
import { executeSearch } from '../../../features/search/search-actions.js';
import { searchReducer as search } from '../../../features/search/search-slice.js';
import { loadReducerError } from '../../../utils/errors.js';
import { buildController, } from '../../controller/headless-controller.js';
import { determineFacetId } from '../../core/facets/_common/facet-id-determinor.js';
import { buildFacetSearch } from '../../core/facets/facet-search/specific/headless-facet-search.js';
/**
 * Creates a `FieldSuggestions` controller instance.
 *
 * This controller initializes a facet under the hood, but exposes state and methods that are relevant for suggesting field values based on a query.
 * It's important not to initialize a facet with the same `facetId` but different options, because only the options of the controller which is built first will be taken into account.
 *
 * @param engine The headless engine.
 * @param props The configurable `FieldSuggestions` controller properties.
 * @returns A `FieldSuggestions` controller instance.
 *
 * @group Controllers
 * @category FieldSuggestions
 */
export function buildFieldSuggestions(engine, props) {
    if (!loadFieldSuggestionsReducers(engine)) {
        throw loadReducerError;
    }
    const { facetSearch: facetSearchOptions, allowedValues, ...facetOptions } = props.options.facet;
    const facetId = determineFacetId(engine, facetOptions);
    engine.dispatch(registerFacet({
        ...defaultFacetOptions,
        ...facetOptions,
        facetId,
        ...(allowedValues && {
            allowedValues: {
                type: 'simple',
                values: allowedValues,
            },
        }),
    }));
    const facetSearch = buildFacetSearch(engine, {
        options: { ...facetSearchOptions, facetId },
        select: (value) => {
            engine.dispatch(updateFacetOptions());
            engine.dispatch(executeSearch({
                legacy: logFacetSelect({ facetId, facetValue: value.rawValue }),
                next: facetSelect(),
            }));
        },
        exclude: (value) => {
            engine.dispatch(updateFacetOptions());
            engine.dispatch(executeSearch({
                legacy: logFacetExclude({ facetId, facetValue: value.rawValue }),
                next: facetExclude(),
            }));
        },
        isForFieldSuggestions: true,
        executeFacetSearchActionCreator: executeFacetSearch,
        executeFieldSuggestActionCreator: executeFieldSuggest,
    });
    const controller = buildController(engine);
    return {
        ...controller,
        ...facetSearch,
        updateText: (text) => {
            facetSearch.updateText(text);
            facetSearch.search();
        },
        get state() {
            return facetSearch.state;
        },
    };
}
function loadFieldSuggestionsReducers(engine) {
    engine.addReducers({ facetSet, configuration, facetSearchSet, search });
    return true;
}
