// @coveo/headless/dist/esm/features/result/result-analytics-actions.js
import { documentIdentifier, makeAnalyticsAction, partialDocumentInformation, validateResultPayload, } from '../analytics/analytics-utils.js';
export const logDocumentOpen = (result) => makeAnalyticsAction({
    prefix: 'analytics/result/open',
    __legacy__getBuilder: (client, state) => {
        validateResultPayload(result);
        return client.makeDocumentOpen(partialDocumentInformation(result, state), documentIdentifier(result));
    },
    analyticsType: 'itemClick',
    analyticsPayloadBuilder: (state) => {
        const docInfo = partialDocumentInformation(result, state);
        const docId = documentIdentifier(result);
        return {
            responseId: result.searchUid ?? '',
            position: docInfo.documentPosition,
            itemMetadata: {
                uniqueFieldName: docId.contentIDKey,
                uniqueFieldValue: docId.contentIDValue,
                title: docInfo.documentTitle,
                author: docInfo.documentAuthor,
                url: docInfo.documentUrl,
            },
        };
    },
});
