// @coveo/headless/dist/esm/features/facets/facet-set/facet-set-insight-analytics-actions.js
import { Value } from '@coveo/bueno';
import { requiredNonEmptyString, validatePayload, } from '../../../utils/validate-payload.js';
import { makeInsightAnalyticsActionFactory, } from '../../analytics/analytics-utils.js';
import { SearchPageEvents } from '../../analytics/search-action-cause.js';
import { getCaseContextAnalyticsMetadata } from '../../case-context/case-context-state.js';
import { facetIdDefinition } from '../generic/facet-actions-validation.js';
import { buildFacetBaseMetadata, buildFacetSelectionChangeMetadata, getStateNeededForFacetMetadata, } from './facet-set-analytics-actions-utils.js';
export const logFacetShowMore = (facetId) => makeInsightAnalyticsActionFactory(SearchPageEvents.facetShowMore)('analytics/facet/showMore', (client, state) => {
    validatePayload(facetId, facetIdDefinition);
    const metadata = {
        ...buildFacetBaseMetadata(facetId, getStateNeededForFacetMetadata(state)),
        ...getCaseContextAnalyticsMetadata(state.insightCaseContext),
    };
    return client.logFacetShowMore(metadata);
});
export const logFacetShowLess = (facetId) => makeInsightAnalyticsActionFactory(SearchPageEvents.facetShowLess)('analytics/facet/showLess', (client, state) => {
    validatePayload(facetId, facetIdDefinition);
    const metadata = {
        ...buildFacetBaseMetadata(facetId, getStateNeededForFacetMetadata(state)),
        ...getCaseContextAnalyticsMetadata(state.insightCaseContext),
    };
    return client.logFacetShowLess(metadata);
});
export const logFacetUpdateSort = (payload) => makeInsightAnalyticsActionFactory(SearchPageEvents.facetUpdateSort)('analytics/facet/sortChange', (client, state) => {
    validatePayload(payload, {
        facetId: facetIdDefinition,
        sortCriterion: new Value({
            required: true,
        }),
    });
    const { facetId, sortCriterion } = payload;
    const stateForAnalytics = getStateNeededForFacetMetadata(state);
    const base = buildFacetBaseMetadata(facetId, stateForAnalytics);
    const metadata = {
        ...base,
        criteria: sortCriterion,
        ...getCaseContextAnalyticsMetadata(state.insightCaseContext),
    };
    return client.logFacetUpdateSort(metadata);
});
export const logFacetClearAll = (facetId) => makeInsightAnalyticsActionFactory(SearchPageEvents.facetClearAll)('analytics/facet/reset', (client, state) => {
    validatePayload(facetId, facetIdDefinition);
    const stateForAnalytics = getStateNeededForFacetMetadata(state);
    const metadata = {
        ...buildFacetBaseMetadata(facetId, stateForAnalytics),
        ...getCaseContextAnalyticsMetadata(state.insightCaseContext),
    };
    return client.logFacetClearAll(metadata);
});
export const logFacetSelect = (payload) => makeInsightAnalyticsActionFactory(SearchPageEvents.facetSelect)('analytics/facet/select', (client, state) => {
    validatePayload(payload, {
        facetId: facetIdDefinition,
        facetValue: requiredNonEmptyString,
    });
    const stateForAnalytics = getStateNeededForFacetMetadata(state);
    const metadata = {
        ...buildFacetSelectionChangeMetadata(payload, stateForAnalytics),
        ...getCaseContextAnalyticsMetadata(state.insightCaseContext),
    };
    return client.logFacetSelect(metadata);
});
export const logFacetDeselect = (payload) => makeInsightAnalyticsActionFactory(SearchPageEvents.facetDeselect)('analytics/facet/deselect', (client, state) => {
    validatePayload(payload, {
        facetId: facetIdDefinition,
        facetValue: requiredNonEmptyString,
    });
    const stateForAnalytics = getStateNeededForFacetMetadata(state);
    const metadata = {
        ...buildFacetSelectionChangeMetadata(payload, stateForAnalytics),
        ...getCaseContextAnalyticsMetadata(state.insightCaseContext),
    };
    return client.logFacetDeselect(metadata);
});
export const logFacetBreadcrumb = (payload) => makeInsightAnalyticsActionFactory(SearchPageEvents.breadcrumbFacet)('analytics/facet/breadcrumb', (client, state) => {
    validatePayload(payload, {
        facetId: facetIdDefinition,
        facetValue: requiredNonEmptyString,
    });
    const metadata = {
        ...buildFacetSelectionChangeMetadata(payload, getStateNeededForFacetMetadata(state)),
        ...getCaseContextAnalyticsMetadata(state.insightCaseContext),
    };
    return client.logBreadcrumbFacet(metadata);
});
