// @coveo/headless/dist/esm/features/facets/range-facets/date-facet-set/date-facet-selectors.js
import { baseFacetResponseSelector } from '../../facet-set/facet-set-selectors.js';
function isDateFacetResponse(state, response) {
    return !!response && response.facetId in state.dateFacetSet;
}
export const dateFacetResponseSelector = (state, facetId) => {
    const response = baseFacetResponseSelector(state, facetId);
    if (isDateFacetResponse(state, response)) {
        return response;
    }
    return undefined;
};
export const dateFacetSelectedValuesSelector = (state, facetId) => {
    const facetResponse = dateFacetResponseSelector(state, facetId) || {
        values: [],
    };
    return facetResponse.values.filter((value) => value.state === 'selected');
};
export const dateFacetActiveValuesSelector = (state, facetId) => {
    const facetResponse = dateFacetResponseSelector(state, facetId) || {
        values: [],
    };
    return facetResponse.values.filter((value) => value.state !== 'idle');
};
export const dateFacetExcludedValuesSelector = (state, facetId) => {
    const facetResponse = dateFacetResponseSelector(state, facetId) || {
        values: [],
    };
    return facetResponse.values.filter((value) => value.state === 'excluded');
};
