// @coveo/headless/dist/esm/api/knowledge/answer-generation/endpoints/follow-up/follow-up-endpoint.js
import { answerGenerationApi } from '../../answer-generation-api.js';
import { initialAnswerGenerationServerState, } from '../../answer-generation-api-state.js';
import { streamAnswerWithStrategy } from '../../streaming/answer-streaming-runner.js';
import { streamingStrategyCreators } from '../../streaming/strategies/streaming-strategy-creators.js';
import { buildFollowUpEndpointUrl } from './url-builders/endpoint-url-builder.js';
/**
 * RTK Query endpoint for streaming follow-up answer generation.
 */
export const followUpEndpoint = answerGenerationApi.injectEndpoints({
    overrideExisting: true,
    endpoints: (builder) => ({
        generateFollowUpAnswer: builder.query({
            queryFn: () => {
                return {
                    data: initialAnswerGenerationServerState(),
                };
            },
            async onQueryStarted(args, { getState, updateCachedData, dispatch }) {
                const { strategyKey, ...params } = args;
                const endpointUrl = buildFollowUpEndpointUrl(getState());
                await streamAnswerWithStrategy(endpointUrl, params, {
                    getState: getState,
                    updateCachedData,
                    dispatch,
                }, streamingStrategyCreators[strategyKey]?.());
            },
        }),
    }),
});
/**
 * Initiates a follow up answer generation query with the specified strategy and parameters.
 */
export const initiateFollowUpEndpoint = (args) => {
    return followUpEndpoint.endpoints.generateFollowUpAnswer.initiate(args);
};
