// @coveo/headless/dist/esm/ssr-next/search/factories/build-factory.js
import { buildSearchEngine, } from '../../../app/search-engine/search-engine.js';
import { loadConfigurationActions } from '../../../features/configuration/configuration-actions-loader.js';
import { createWaitForActionMiddleware } from '../../../utils/utils.js';
import { buildControllerDefinitions } from '../controller-utils.js';
import { wireControllerParams } from '../utils/controller-wiring.js';
import { augmentSearchEngineOptions } from '../utils/engine-wiring.js';
function isSearchCompletedAction(action) {
    return /^search\/executeSearch\/(fulfilled|rejected)$/.test(action.type);
}
function buildSSRSearchEngine(options) {
    const { middleware, promise } = createWaitForActionMiddleware(isSearchCompletedAction);
    const searchEngine = buildSearchEngine({
        ...options,
        middlewares: [...(options.middlewares ?? []), middleware],
    });
    return {
        ...searchEngine,
        get state() {
            return searchEngine.state;
        },
        waitForSearchCompletedAction() {
            return promise;
        },
    };
}
export const buildFactory = (controllerDefinitions, options) => async (buildOptions) => {
    const controllerProps = wireControllerParams(controllerDefinitions, buildOptions);
    const engineOptions = augmentSearchEngineOptions(options, buildOptions);
    const engine = buildSSRSearchEngine(engineOptions);
    const updateEngineConfiguration = (accessToken) => {
        const { updateBasicConfiguration } = loadConfigurationActions(engine);
        engine.dispatch(updateBasicConfiguration({
            accessToken,
        }));
    };
    if (options.onAccessTokenUpdate) {
        options.onAccessTokenUpdate(updateEngineConfiguration);
    }
    const controllers = buildControllerDefinitions({
        definitionsMap: controllerDefinitions,
        engine,
        propsMap: controllerProps,
    });
    if (buildOptions && 'searchActions' in buildOptions) {
        buildOptions.searchActions.forEach((action) => {
            engine.dispatch(action);
        });
    }
    return {
        engine,
        controllers: controllers,
    };
};
