// @coveo/headless/dist/esm/ssr-next/common/builders/controller-builder.js
/**
 * Builder class for creating controllers from definitions
 */
export class ControllerBuilder {
    _definition;
    _engine;
    _props;
    _additionalArgs = [];
    constructor(definition, engine, props) {
        this._definition = definition;
        this._engine = engine;
        this._props = props;
    }
    withAdditionalArgs(additionalArgs = []) {
        this._additionalArgs = additionalArgs;
        return this;
    }
    build() {
        if ('build' in this._definition &&
            typeof this._definition.build === 'function') {
            return this.buildWithoutProps();
        }
        if ('buildWithProps' in this._definition &&
            typeof this._definition.buildWithProps === 'function') {
            return this.buildWithProps();
        }
        throw new Error('Controller definition must have a build or buildWithProps method.');
    }
    buildWithoutProps() {
        const buildFn = this._definition.build;
        const controller = buildFn(this._engine, ...this._additionalArgs);
        return controller;
    }
    buildWithProps() {
        const buildWithPropsFn = this._definition.buildWithProps;
        const controller = buildWithPropsFn(this._engine, this._props, ...this._additionalArgs);
        const controllerWithInitialState = Object.create(Object.getPrototypeOf(controller), {
            ...Object.getOwnPropertyDescriptors(controller),
            initialState: {
                value: controller.state,
                writable: false,
                enumerable: true,
                configurable: false,
            },
        });
        controllerWithInitialState.subscribe =
            controller.subscribe.bind(controller);
        return controllerWithInitialState;
    }
}
