// @coveo/headless/dist/esm/features/document-suggestion/document-suggestion-actions.js
import { createAsyncThunk } from '@reduxjs/toolkit';
import { getVisitorID } from '../../api/analytics/coveo-analytics-utils.js';
import { getOrganizationEndpoint } from '../../api/platform-client.js';
import { isErrorResponse } from '../../api/search/search-api-client.js';
import { prepareContextFromFields } from '../../api/service/case-assist/case-assist-params.js';
import { SearchPageEvents } from '../analytics/search-action-cause.js';
import { fromAnalyticsStateToAnalyticsParams } from '../configuration/analytics-params.js';
export const fetchDocumentSuggestions = createAsyncThunk('caseAssist/documentSuggestions/fetch', async (_, { getState, rejectWithValue, extra: { apiClient, navigatorContext } }) => {
    const state = getState();
    const fetched = await apiClient.getDocumentSuggestions(await buildFetchDocumentSuggestionsRequest(state, navigatorContext));
    if (isErrorResponse(fetched)) {
        return rejectWithValue(fetched.error);
    }
    return {
        response: fetched.success,
    };
});
export const buildFetchDocumentSuggestionsRequest = async (state, navigatorContext) => ({
    accessToken: state.configuration.accessToken,
    organizationId: state.configuration.organizationId,
    url: state.caseAssistConfiguration.apiBaseUrl ??
        getOrganizationEndpoint(state.configuration.organizationId, state.configuration.environment),
    caseAssistId: state.caseAssistConfiguration.caseAssistId,
    ...(state.configuration.analytics.enabled && {
        clientId: await getVisitorID(state.configuration.analytics),
    }),
    ...(state.configuration.analytics.enabled &&
        fromAnalyticsStateToAnalyticsParams(state.configuration.analytics, navigatorContext, {
            actionCause: SearchPageEvents.documentSuggestion,
        })),
    fields: state.caseInput,
    context: state.caseField
        ? prepareContextFromFields(state.caseField.fields)
        : undefined,
    locale: state.caseAssistConfiguration.locale,
    debug: state.debug,
});
