// @coveo/headless/dist/esm/controllers/recommendation/headless-recommendation.js
import { NumberValue, Schema, StringValue } from '@coveo/bueno';
import { configuration } from '../../app/common-reducers.js';
import { loadPaginationActions } from '../../features/pagination/pagination-actions-loader.js';
import { getRecommendations, setRecommendationId, } from '../../features/recommendation/recommendation-actions.js';
import { recommendationReducer as recommendation } from '../../features/recommendation/recommendation-slice.js';
import { loadReducerError } from '../../utils/errors.js';
import { validateOptions } from '../../utils/validate-payload.js';
import { buildController, } from '../controller/headless-controller.js';
const optionsSchema = new Schema({
    id: new StringValue({
        emptyAllowed: true,
        required: false,
        default: '',
    }),
    numberOfRecommendations: new NumberValue({ min: 0 }),
});
/**
 * Creates a `RecommendationList` controller instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `RecommendationList` properties.
 * @returns A `RecommendationList` controller instance.
 *
 * @group Controllers
 * @category RecommendationList
 */
export function buildRecommendationList(engine, props = {}) {
    if (!loadRecommendationListReducers(engine)) {
        throw loadReducerError;
    }
    const controller = buildController(engine);
    const { dispatch } = engine;
    const getState = () => engine.state;
    const options = validateOptions(engine, optionsSchema, props.options, 'buildRecommendationList');
    if (options.id !== '') {
        dispatch(setRecommendationId({ id: options.id }));
    }
    if (options.numberOfRecommendations) {
        dispatch(loadPaginationActions(engine).updateNumberOfResults(options.numberOfRecommendations));
    }
    return {
        ...controller,
        refresh() {
            dispatch(getRecommendations());
        },
        get state() {
            const state = getState();
            return {
                recommendations: state.recommendation.recommendations,
                error: state.recommendation.error,
                isLoading: state.recommendation.isLoading,
                searchResponseId: state.recommendation.searchUid,
            };
        },
    };
}
function loadRecommendationListReducers(engine) {
    engine.addReducers({ recommendation, configuration });
    return true;
}
