// @coveo/headless/dist/esm/controllers/standalone-search-box/headless-standalone-search-box.js
import { configuration } from '../../app/common-reducers.js';
import { updateQuery } from '../../features/query/query-actions.js';
import { queryReducer as query } from '../../features/query/query-slice.js';
import { selectQuerySuggestion } from '../../features/query-suggest/query-suggest-actions.js';
import { buildOmniboxSuggestionMetadata } from '../../features/query-suggest/query-suggest-analytics-actions.js';
import { querySuggestReducer as querySuggest } from '../../features/query-suggest/query-suggest-slice.js';
import { fetchRedirectUrl, registerStandaloneSearchBox, resetStandaloneSearchBox, updateAnalyticsToOmniboxFromLink, updateAnalyticsToSearchFromLink, updateStandaloneSearchBoxRedirectionUrl, } from '../../features/standalone-search-box-set/standalone-search-box-set-actions.js';
import { standaloneSearchBoxSetReducer as standaloneSearchBoxSet } from '../../features/standalone-search-box-set/standalone-search-box-set-slice.js';
import { loadReducerError } from '../../utils/errors.js';
import { randomID } from '../../utils/utils.js';
import { validateOptions } from '../../utils/validate-payload.js';
import { defaultSearchBoxOptions } from '../core/search-box/headless-core-search-box-options.js';
import { buildSearchBox, } from '../search-box/headless-search-box.js';
import { standaloneSearchBoxSchema, } from './headless-standalone-search-box-options.js';
/**
 * Creates a `StandaloneSearchBox` instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `StandaloneSearchBox` properties.
 * @returns A `StandaloneSearchBox` instance.
 *
 * @group Controllers
 * @category StandaloneSearchBox
 */
export function buildStandaloneSearchBox(engine, props) {
    if (!loadStandaloneSearchBoxReducers(engine)) {
        throw loadReducerError;
    }
    const { dispatch } = engine;
    const getState = () => engine.state;
    const id = props.options.id || randomID('standalone_search_box');
    const options = {
        id,
        highlightOptions: { ...props.options.highlightOptions },
        ...defaultSearchBoxOptions,
        ...{ overwrite: false },
        ...props.options,
    };
    validateOptions(engine, standaloneSearchBoxSchema, options, 'buildStandaloneSearchBox');
    const searchBox = buildSearchBox(engine, { options });
    dispatch(registerStandaloneSearchBox({
        id,
        redirectionUrl: options.redirectionUrl,
        overwrite: options.overwrite,
    }));
    return {
        ...searchBox,
        updateText(value) {
            searchBox.updateText(value);
            dispatch(updateAnalyticsToSearchFromLink({ id }));
        },
        selectSuggestion(value) {
            const metadata = buildOmniboxSuggestionMetadata(getState(), {
                id,
                suggestion: value,
            });
            dispatch(selectQuerySuggestion({ id, expression: value }));
            dispatch(updateAnalyticsToOmniboxFromLink({ id, metadata }));
            this.submit();
        },
        afterRedirection() {
            dispatch(resetStandaloneSearchBox({ id }));
        },
        updateRedirectUrl(url) {
            dispatch(updateStandaloneSearchBoxRedirectionUrl({ id, redirectionUrl: url }));
        },
        submit() {
            dispatch(updateQuery({
                q: this.state.value,
                enableQuerySyntax: options.enableQuerySyntax,
            }));
            dispatch(fetchRedirectUrl({ id }));
        },
        get state() {
            const state = getState();
            const standaloneSearchBoxState = state.standaloneSearchBoxSet[id];
            return {
                ...searchBox.state,
                isLoading: standaloneSearchBoxState.isLoading,
                redirectTo: standaloneSearchBoxState.redirectTo,
                analytics: standaloneSearchBoxState.analytics,
            };
        },
    };
}
function loadStandaloneSearchBoxReducers(engine) {
    engine.addReducers({
        standaloneSearchBoxSet,
        configuration,
        query,
        querySuggest,
    });
    return true;
}
