// @coveo/headless/dist/esm/features/pagination/pagination-analytics-actions.js
import { makeAnalyticsAction, } from '../analytics/analytics-utils.js';
import { SearchPageEvents } from '../analytics/search-action-cause.js';
import { currentPageSelector } from './pagination-selectors.js';
import { getPaginationInitialState } from './pagination-state.js';
export const logPagerResize = () => makeAnalyticsAction('analytics/pager/resize', (client, state) => client.makePagerResize({
    currentResultsPerPage: state.pagination?.numberOfResults ||
        getPaginationInitialState().numberOfResults,
}));
export const logPageNumber = () => makeAnalyticsAction('analytics/pager/number', (client, state) => client.makePagerNumber({
    pagerNumber: currentPageSelector(state),
}));
export const logPageNext = () => makeAnalyticsAction('analytics/pager/next', (client, state) => client.makePagerNext({
    pagerNumber: currentPageSelector(state),
}));
export const logPagePrevious = () => makeAnalyticsAction('analytics/pager/previous', (client, state) => client.makePagerPrevious({
    pagerNumber: currentPageSelector(state),
}));
// TODO KIT-2983
export const browseResults = () => ({
    actionCause: SearchPageEvents.browseResults,
});
