// @coveo/headless/dist/esm/controllers/facets/range-facet/numeric-facet/headless-numeric-facet.js
import { configuration } from '../../../../app/common-reducers.js';
import { facetClearAll, logFacetClearAll, logFacetUpdateSort, } from '../../../../features/facets/facet-set/facet-set-analytics-actions.js';
import { getAnalyticsActionForToggleFacetSelect, getLegacyAnalyticsActionForToggleRangeFacetSelect, } from '../../../../features/facets/range-facets/generic/range-facet-utils.js';
import { numericFacetSetReducer as numericFacetSet } from '../../../../features/facets/range-facets/numeric-facet-set/numeric-facet-set-slice.js';
import { executeSearch } from '../../../../features/search/search-actions.js';
import { searchReducer as search } from '../../../../features/search/search-slice.js';
import { loadReducerError } from '../../../../utils/errors.js';
import { buildCoreNumericFacet, buildNumericRange, } from '../../../core/facets/range-facet/numeric-facet/headless-core-numeric-facet.js';
export { buildNumericRange };
/**
 * Creates a `NumericFacet` controller instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `NumericFacet` properties.
 * @returns A `NumericFacet` controller instance.
 *
 * @group Controllers
 * @category NumericFacet
 */
export function buildNumericFacet(engine, props) {
    if (!loadNumericFacetReducers(engine)) {
        throw loadReducerError;
    }
    const coreController = buildCoreNumericFacet(engine, props);
    const dispatch = engine.dispatch;
    const getFacetId = () => coreController.state.facetId;
    return {
        ...coreController,
        deselectAll() {
            coreController.deselectAll();
            dispatch(executeSearch({
                legacy: logFacetClearAll(getFacetId()),
                next: facetClearAll(),
            }));
        },
        sortBy(criterion) {
            coreController.sortBy(criterion);
            dispatch(executeSearch({
                legacy: logFacetUpdateSort({ facetId: getFacetId(), criterion }),
            }));
        },
        toggleSelect: (selection) => {
            coreController.toggleSelect(selection);
            dispatch(executeSearch({
                legacy: getLegacyAnalyticsActionForToggleRangeFacetSelect(getFacetId(), selection),
                next: getAnalyticsActionForToggleFacetSelect(selection),
            }));
        },
        get state() {
            return {
                ...coreController.state,
            };
        },
    };
}
function loadNumericFacetReducers(engine) {
    engine.addReducers({ numericFacetSet, configuration, search });
    return true;
}
