// @coveo/headless/dist/esm/case-assist.index.js
/**
 * The Coveo Headless Case Assist sub-package exposes exposes the engine, controllers, actions, and utility functions to build a case assist experience.
 *
 * @example
 * ```typescript
 * import { buildCaseInput, buildCaseAssistEngine } from '@coveo/headless/case-assist';
 *
 * const engine = buildCaseAssistEngine({
 *    configuration: ...
 * });
 *
 * const caseInput = buildCaseInput(engine);
 * ```
 * @module Case Assist
 */
export { getAnalyticsNextApiBaseUrl, getOrganizationEndpoint, } from './api/platform-client.js';
export { buildCaseAssistEngine } from './app/case-assist-engine/case-assist-engine.js';
export { buildCaseField } from './controllers/case-field/headless-case-field.js';
export { buildCaseInput } from './controllers/case-input/headless-case-input.js';
export { buildController } from './controllers/controller/headless-controller.js';
export { buildCaseAssistInteractiveResult as buildInteractiveResult } from './controllers/document-suggestion-list/case-assist-headless-interactive-result.js';
export { buildDocumentSuggestionList, buildDocumentSuggestionList as buildDocumentSuggestion, } from './controllers/document-suggestion-list/headless-document-suggestion-list.js';
export { buildCaseAssistQuickview as buildQuickview } from './controllers/quickview/case-assist-headless-quickview.js';
export { loadGenericAnalyticsActions } from './features/analytics/generic-analytics-actions-loader.js';
export { loadCaseAssistAnalyticsActions } from './features/case-assist/case-assist-analytics-actions-loader.js';
export { loadCaseFieldActions } from './features/case-field/case-field-actions-loader.js';
// Case Assist Action Loaders
export { loadCaseInputActions } from './features/case-input/case-input-actions-loader.js';
export { loadDocumentSuggestionActions } from './features/document-suggestion/document-suggestion-actions-loader.js';
// Features
export { ResultTemplatesHelpers } from './features/result-templates/result-templates-helpers.js';
