// @coveo/headless/dist/esm/insight.index.js
/**
 * The Coveo Headless Insight sub-package exposes exposes the engine, controllers, actions, and utility functions to build an insight experience.
 *
 * @example
 * ```typescript
 * import { buildSearchBox, buildInsightEngine, getSampleInsightEngineConfiguration } from '@coveo/headless/insight';
 *
 * const engine = buildInsightEngine({
 *    configuration: getSampleInsightEngineConfiguration()
 * });
 *
 * const searchBox = buildSearchBox(engine);
 * ```
 * @module Insight
 */
import * as HighlightUtils from './utils/highlight.js';
export { buildInsightEngine, getSampleInsightEngineConfiguration, } from './app/insight-engine/insight-engine.js';
export { buildController } from './controllers/controller/headless-controller.js';
export { buildAttachToCase } from './controllers/insight/attach-to-case/headless-attach-to-case.js';
export { buildAttachedResults } from './controllers/insight/attached-results/headless-attached-results.js';
export { buildBreadcrumbManager } from './controllers/insight/breadcrumb-manager/headless-insight-breadcrumb-manager.js';
export { buildDidYouMean } from './controllers/insight/did-you-mean/headless-insight-did-you-mean.js';
export { buildFacetConditionsManager } from './controllers/insight/facet-conditions-manager/headless-facet-conditions-manager.js';
export { buildFacetManager } from './controllers/insight/facet-manager/headless-insight-facet-manager.js';
export { buildCategoryFacet } from './controllers/insight/facets/category-facet/headless-insight-category-facet.js';
export { buildFacet } from './controllers/insight/facets/facet/headless-insight-facet.js';
export { buildDateFacet, buildDateRange, } from './controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.js';
export { buildDateFilter } from './controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.js';
export { buildNumericFacet, buildNumericRange, } from './controllers/insight/facets/range-facet/numeric-facet/headless-insight-numeric-facet.js';
export { buildNumericFilter } from './controllers/insight/facets/range-facet/numeric-facet/headless-insight-numeric-filter.js';
export { buildFoldedResultList } from './controllers/insight/folded-result-list/headless-insight-folded-result-list.js';
export { buildGeneratedAnswer } from './controllers/insight/generated-answer/headless-insight-generated-answer.js';
export { buildInteractiveCitation } from './controllers/insight/generated-answer/headless-insight-interactive-citation.js';
export { buildPager } from './controllers/insight/pager/headless-insight-pager.js';
export { buildQueryError } from './controllers/insight/query-error/headless-insight-query-error.js';
export { buildQuerySummary } from './controllers/insight/query-summary/headless-insight-query-summary.js';
export { buildQuickview } from './controllers/insight/quickview/headless-insight-quickview.js';
export { buildRecentQueriesList } from './controllers/insight/recent-queries-list/headless-insight-recent-queries-list.js';
export { buildInteractiveResult } from './controllers/insight/result-list/headless-insight-interactive-result.js';
export { buildResultList } from './controllers/insight/result-list/headless-insight-result-list.js';
export { buildResultsPerPage } from './controllers/insight/results-per-page/headless-insight-results-per-page.js';
export { buildSearchBox } from './controllers/insight/search-box/headless-insight-search-box.js';
export { buildSearchParameterManager } from './controllers/insight/search-parameter-manager/headless-insight-search-parameter-manager.js';
export { buildSmartSnippet } from './controllers/insight/smart-snippet/headless-insight-smart-snippet.js';
export { buildSmartSnippetQuestionsList } from './controllers/insight/smart-snippet-questions-list/headless-insight-smart-snippet-questions-list.js';
export { buildSort } from './controllers/insight/sort/headless-insight-sort.js';
export { buildSearchStatus } from './controllers/insight/status/headless-insight-status.js';
export { buildTab } from './controllers/insight/tab/headless-insight-tab.js';
export { buildNotifyTrigger } from './controllers/insight/triggers/headless-insight-notify-trigger.js';
export { buildUserActions } from './controllers/insight/user-actions/headless-user-actions.js';
export { buildInsightInterface } from './controllers/insight-interface/insight-interface.js';
export { loadGenericAnalyticsActions } from './features/analytics/generic-analytics-actions-loader.js';
export { loadInsightAnalyticsActions } from './features/analytics/insight-analytics-actions-loader.js';
export { loadAttachedResultsActions } from './features/attached-results/attached-results-actions-loader.js';
export { loadAttachedResultsAnalyticsActions } from './features/attached-results/attached-results-analytics-actions-loader.js';
export { loadCaseContextActions } from './features/case-context/case-context-actions-loader.js';
export { loadContextActions } from './features/context/context-actions-loader.js';
export { loadDateFacetSetActions } from './features/facets/range-facets/date-facet-set/date-facet-actions-loader.js';
export { loadNumericFacetSetActions } from './features/facets/range-facets/numeric-facet-set/numeric-facet-actions-loader.js';
export { loadFieldActions } from './features/fields/fields-actions-loader.js';
export { loadFoldingActions } from './features/folding/insight-folding-actions-loader.js';
export { loadGeneratedAnswerActions } from './features/generated-answer/generated-answer-actions-loader.js';
// Action loaders
export { loadInsightInterfaceActions } from './features/insight-interface/insight-interface-actions-loader.js';
export { loadQuerySetActions } from './features/insight-search/insight-query-set-actions-loader.js';
export { loadInsightSearchActions } from './features/insight-search/insight-search-actions-loader.js';
export { loadInsightSearchAnalyticsActions } from './features/insight-search/insight-search-analytics-actions-loader.js';
export { loadInsightUserActionsActions } from './features/insight-user-actions/insight-user-actions-loader.js';
export { loadQuerySuggestActions } from './features/query-suggest/query-suggest-actions-loader.js';
export { loadQuestionAnsweringActions } from './features/question-answering/question-answering-actions-loader.js';
export { loadRecentResultsActions } from './features/recent-results/recent-results-actions-loader.js';
export { ResultTemplatesHelpers } from './features/result-templates/result-templates-helpers.js';
export { buildResultTemplatesManager } from './features/result-templates/result-templates-manager.js';
export { HighlightUtils };
export { getAnalyticsNextApiBaseUrl, getOrganizationEndpoint, } from './api/platform-client.js';
export { deserializeRelativeDate } from './api/search/date/relative-date.js';
export { buildCriterionExpression, buildDateSortCriterion, buildRelevanceSortCriterion, SortOrder, } from './features/sort-criteria/criteria.js';
export { VERSION } from './utils/version.js';
