// @coveo/headless/dist/esm/recommendation.index.js
/**
 * The Coveo Headless Recommendation sub-package exposes exposes the engine, controllers, actions, and utility functions to build a recommendation experience.
 *
 * @example
 * ```typescript
 * import { buildRecommendationList, buildRecommendationEngine, getSampleRecommendationEngineConfiguration } from '@coveo/headless/recommendation';
 *
 * const engine = buildRecommendationEngine({
 *    configuration: getSampleRecommendationEngineConfiguration()
 * });
 *
 * const recommendationList = buildRecommendationList(engine);
 * ```
 * @module Recommendation
 */
import * as HighlightUtils from './utils/highlight.js';
export { HighlightUtils };
export { getAnalyticsNextApiBaseUrl, getOrganizationEndpoint, } from './api/platform-client.js';
export { buildRecommendationEngine, getSampleRecommendationEngineConfiguration, } from './app/recommendation-engine/recommendation-engine.js';
export { buildContext } from './controllers/context/headless-context.js';
export { buildController } from './controllers/controller/headless-controller.js';
export { buildDictionaryFieldContext } from './controllers/dictionary-field-context/headless-dictionary-field-context.js';
export { buildRecommendationList } from './controllers/recommendation/headless-recommendation.js';
export { buildInteractiveResult } from './controllers/recommendation/result-list/headless-recommendation-interactive-result.js';
export { loadAdvancedSearchQueryActions } from './features/advanced-search-queries/advanced-search-queries-actions-loader.js';
// Actions
export { loadConfigurationActions } from './features/configuration/configuration-actions-loader.js';
export { loadSearchConfigurationActions } from './features/configuration/search-configuration-actions-loader.js';
export { loadContextActions } from './features/context/context-actions-loader.js';
export { loadDebugActions } from './features/debug/debug-actions-loader.js';
export { loadDictionaryFieldContextActions } from './features/dictionary-field-context/dictionary-field-context-actions-loader.js';
export { loadFieldActions } from './features/fields/fields-actions-loader.js';
export { DefaultFieldsToInclude, EcommerceDefaultFieldsToInclude, MinimumFieldsToInclude, } from './features/fields/fields-state.js';
export { loadPaginationActions } from './features/pagination/pagination-actions-loader.js';
export { loadPipelineActions } from './features/pipeline/pipeline-actions-loader.js';
export { loadRecommendationActions } from './features/recommendation/recommendation-actions-loader.js';
export { loadClickAnalyticsActions } from './features/recommendation/recommendation-click-analytics-actions-loader.js';
export { ResultTemplatesHelpers } from './features/result-templates/result-templates-helpers.js';
export { buildResultTemplatesManager } from './features/result-templates/result-templates-manager.js';
export { loadSearchHubActions } from './features/search-hub/search-hub-actions-loader.js';
