// @coveo/headless/dist/esm/api/service/insight/insight-api-client.js
import { PlatformClient } from '../../platform-client.js';
import { getHtml, } from '../../search/html/html-api-client.js';
import { isSuccessSearchResponse, shimResponse, } from '../../search/search-api-client.js';
import { buildAPIResponseFromErrorOrThrow } from '../../search/search-api-error-response.js';
import { buildGetInsightInterfaceConfigRequest, } from './get-interface/get-interface-config-request.js';
import { buildInsightQueryRequest, buildInsightQuerySuggestRequest, } from './query/query-request.js';
import { buildInsightUserActionsRequest, } from './user-actions/user-actions-request.js';
/**
 * The client to use to interact with the Insight API.
 */
export class InsightAPIClient {
    options;
    constructor(options) {
        this.options = options;
    }
    async getInterface(req) {
        const response = await PlatformClient.call({
            ...buildGetInsightInterfaceConfigRequest(req),
            ...this.options,
        });
        if (response instanceof Error) {
            return buildAPIResponseFromErrorOrThrow(response);
        }
        const body = await response.json();
        return response.ok
            ? { success: body }
            : { error: body };
    }
    async query(req, options) {
        const response = await PlatformClient.call({
            ...buildInsightQueryRequest(req),
            requestMetadata: { method: 'search', origin: options?.origin },
            ...this.options,
        });
        if (response instanceof Error) {
            return buildAPIResponseFromErrorOrThrow(response);
        }
        let body = await response.json();
        if (isSuccessSearchResponse(body)) {
            body = shimResponse(body);
        }
        return response.ok
            ? { success: body }
            : { error: body };
    }
    async querySuggest(req) {
        const response = await PlatformClient.call({
            ...buildInsightQuerySuggestRequest(req),
            ...this.options,
        });
        if (response instanceof Error) {
            return buildAPIResponseFromErrorOrThrow(response);
        }
        const body = await response.json();
        return body.completions
            ? { success: body }
            : { error: body };
    }
    async userActions(req) {
        const response = await PlatformClient.call({
            ...buildInsightUserActionsRequest(req),
            ...this.options,
        });
        if (response instanceof Error) {
            return buildAPIResponseFromErrorOrThrow(response);
        }
        const body = await response.json();
        return response.ok
            ? { success: body }
            : { error: body };
    }
    async html(req) {
        return getHtml(req, { ...this.options });
    }
}
