// @coveo/headless/dist/esm/app/case-assist-engine/case-assist-engine.js
import { isNullOrUndefined } from '@coveo/bueno';
import { NoopPreprocessRequest } from '../../api/preprocess-request.js';
import { CaseAssistAPIClient } from '../../api/service/case-assist/case-assist-api-client.js';
import { setCaseAssistConfiguration } from '../../features/case-assist-configuration/case-assist-configuration-actions.js';
import { caseAssistConfigurationReducer as caseAssistConfiguration } from '../../features/case-assist-configuration/case-assist-configuration-slice.js';
import { debugReducer as debug } from '../../features/debug/debug-slice.js';
import { setSearchHub } from '../../features/search-hub/search-hub-actions.js';
import { searchHubReducer as searchHub } from '../../features/search-hub/search-hub-slice.js';
import { buildEngine, warnIfUsingNextAnalyticsModeForServiceFeature, } from '../engine.js';
import { buildLogger } from '../logger.js';
import { buildThunkExtraArguments } from '../thunk-extra-arguments.js';
import { caseAssistEngineConfigurationSchema, } from './case-assist-engine-configuration.js';
const caseassistEngineReducers = {
    debug,
    caseAssistConfiguration,
    searchHub,
};
/**
 * Creates a case assist engine instance.
 *
 * @param options - The case assist engine options.
 * @returns A case assist engine instance.
 *
 * @group Engine
 */
export function buildCaseAssistEngine(options) {
    const logger = buildLogger(options.loggerOptions);
    validateConfiguration(options.configuration, logger);
    const caseAssistAPIClient = createCaseAssistAPIClient(options.configuration, logger);
    const thunkArguments = {
        ...buildThunkExtraArguments(options.configuration, logger),
        apiClient: caseAssistAPIClient,
    };
    const augmentedOptions = {
        ...options,
        reducers: caseassistEngineReducers,
    };
    const engine = buildEngine(augmentedOptions, thunkArguments);
    warnIfUsingNextAnalyticsModeForServiceFeature(engine.state.configuration.analytics.analyticsMode);
    const { caseAssistId, locale, searchHub, proxyBaseUrl } = options.configuration;
    engine.dispatch(setCaseAssistConfiguration({
        caseAssistId,
        locale,
        proxyBaseUrl,
    }));
    if (!isNullOrUndefined(searchHub)) {
        engine.dispatch(setSearchHub(searchHub));
    }
    return {
        ...engine,
        get state() {
            return engine.state;
        },
    };
}
function validateConfiguration(configuration, logger) {
    try {
        caseAssistEngineConfigurationSchema.validate(configuration);
    }
    catch (error) {
        logger.error(error, 'Case Assist engine configuration error');
        throw error;
    }
}
function createCaseAssistAPIClient(configuration, logger) {
    return new CaseAssistAPIClient({
        logger,
        preprocessRequest: configuration.preprocessRequest || NoopPreprocessRequest,
    });
}
