// @coveo/headless/dist/esm/controllers/core/facets/range-facet/numeric-facet/headless-core-numeric-filter.js
import { configuration } from '../../../../../app/common-reducers.js';
import { disableFacet, enableFacet, updateFacetOptions, } from '../../../../../features/facet-options/facet-options-actions.js';
import { isFacetEnabledSelector } from '../../../../../features/facet-options/facet-options-selectors.js';
import { facetOptionsReducer as facetOptions } from '../../../../../features/facet-options/facet-options-slice.js';
import { isFacetLoadingResponseSelector } from '../../../../../features/facets/facet-set/facet-set-selectors.js';
import { registerNumericFacet, updateNumericFacetValues, } from '../../../../../features/facets/range-facets/numeric-facet-set/numeric-facet-actions.js';
import { numericFacetSelectedValuesSelector } from '../../../../../features/facets/range-facets/numeric-facet-set/numeric-facet-selectors.js';
import { numericFacetSetReducer as numericFacetSet } from '../../../../../features/facets/range-facets/numeric-facet-set/numeric-facet-set-slice.js';
import { searchReducer as search } from '../../../../../features/search/search-slice.js';
import { selectActiveTab } from '../../../../../features/tab-set/tab-set-selectors.js';
import { loadReducerError } from '../../../../../utils/errors.js';
import { buildController, } from '../../../../controller/headless-controller.js';
import { determineFacetId } from '../../_common/facet-id-determinor.js';
import { validateNumericFacetOptions } from './headless-numeric-facet-options.js';
export function buildCoreNumericFilter(engine, props) {
    if (!loadNumericFilterReducer(engine)) {
        throw loadReducerError;
    }
    const controller = buildController(engine);
    const { dispatch } = engine;
    const getState = () => engine.state;
    const facetId = determineFacetId(engine, props.options);
    const tabs = props.options.tabs ?? {};
    const activeTab = selectActiveTab(engine.state.tabSet);
    const options = {
        ...props.options,
        currentValues: props.initialState?.range
            ? [{ ...props.initialState.range, endInclusive: true, state: 'selected' }]
            : [],
        generateAutomaticRanges: false,
        facetId,
        tabs,
        activeTab,
    };
    validateNumericFacetOptions(engine, options);
    dispatch(registerNumericFacet(options));
    const getIsEnabled = () => isFacetEnabledSelector(engine.state, facetId);
    return {
        ...controller,
        clear: () => {
            dispatch(updateNumericFacetValues({
                facetId,
                values: [],
            }));
            dispatch(updateFacetOptions());
        },
        setRange: (range) => {
            const facetValue = {
                ...range,
                state: 'selected',
                numberOfResults: 0,
                endInclusive: true,
            };
            const updateFacetValuesAction = updateNumericFacetValues({
                facetId,
                values: [facetValue],
            });
            if (updateFacetValuesAction.error) {
                return false;
            }
            dispatch(updateFacetValuesAction);
            dispatch(updateFacetOptions());
            return true;
        },
        enable() {
            dispatch(enableFacet(facetId));
        },
        disable() {
            dispatch(disableFacet(facetId));
        },
        get state() {
            const isLoading = isFacetLoadingResponseSelector(getState());
            const enabled = getIsEnabled();
            const selectedRanges = numericFacetSelectedValuesSelector(getState(), facetId);
            const range = selectedRanges.length ? selectedRanges[0] : undefined;
            return {
                facetId,
                isLoading,
                range,
                enabled,
            };
        },
    };
}
function loadNumericFilterReducer(engine) {
    engine.addReducers({ numericFacetSet, facetOptions, configuration, search });
    return true;
}
