// @coveo/headless/dist/esm/controllers/core/facets/range-facet/numeric-facet/headless-core-numeric-facet.js
import { configuration } from '../../../../../app/common-reducers.js';
import { facetOptionsReducer as facetOptions } from '../../../../../features/facet-options/facet-options-slice.js';
import { deselectAllFacetValues } from '../../../../../features/facets/facet-set/facet-set-actions.js';
import { registerNumericFacet, } from '../../../../../features/facets/range-facets/numeric-facet-set/numeric-facet-actions.js';
import { executeToggleNumericFacetSelect } from '../../../../../features/facets/range-facets/numeric-facet-set/numeric-facet-controller-actions.js';
import { numericFacetSetReducer as numericFacetSet } from '../../../../../features/facets/range-facets/numeric-facet-set/numeric-facet-set-slice.js';
import { searchReducer as search } from '../../../../../features/search/search-slice.js';
import { selectActiveTab } from '../../../../../features/tab-set/tab-set-selectors.js';
import { loadReducerError } from '../../../../../utils/errors.js';
import { determineFacetId } from '../../_common/facet-id-determinor.js';
import { assertRangeFacetOptions } from '../core-range-facet-utils.js';
import { buildCoreRangeFacet } from '../headless-core-range-facet.js';
import { validateNumericFacetOptions, } from './headless-numeric-facet-options.js';
import { buildNumericRange } from './numeric-range.js';
export { buildNumericRange };
/**
 * Creates a `NumericFacet` controller instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `NumericFacet` properties.
 * @returns A `NumericFacet` controller instance.
 *
 * @group Controllers
 * @category NumericFacet
 */
export function buildCoreNumericFacet(engine, props) {
    if (!loadNumericFacetReducers(engine)) {
        throw loadReducerError;
    }
    assertRangeFacetOptions(props.options, 'buildNumericFacet');
    const dispatch = engine.dispatch;
    const facetId = determineFacetId(engine, props.options);
    const tabs = props.options.tabs ?? {};
    const activeTab = selectActiveTab(engine.state.tabSet);
    const options = {
        currentValues: [],
        ...props.options,
        facetId,
        tabs,
        activeTab,
    };
    validateNumericFacetOptions(engine, options);
    dispatch(registerNumericFacet(options));
    const rangeFacet = buildCoreRangeFacet(engine, {
        facetId,
        getRequest: () => engine.state.numericFacetSet[facetId].request,
    });
    return {
        ...rangeFacet,
        toggleSelect: (selection) => dispatch(executeToggleNumericFacetSelect({ facetId, selection })),
        toggleSingleSelect(selection) {
            if (selection.state === 'idle') {
                dispatch(deselectAllFacetValues(facetId));
            }
            this.toggleSelect(selection);
        },
        get state() {
            return rangeFacet.state;
        },
    };
}
function loadNumericFacetReducers(engine) {
    engine.addReducers({ numericFacetSet, facetOptions, configuration, search });
    return true;
}
