// @coveo/headless/dist/esm/features/folding/insight-folding-actions.js
import { createAction, createAsyncThunk } from '@reduxjs/toolkit';
import { isErrorResponse } from '../../api/search/search-api-client.js';
import { validatePayload } from '../../utils/validate-payload.js';
import { foldingOptionsSchemaDefinition, } from '../folding/folding-actions.js';
import { fetchFromAPI } from '../insight-search/insight-search-actions.js';
import { buildInsightLoadCollectionRequest } from '../insight-search/insight-search-request.js';
export const registerFolding = createAction('folding/register', (payload) => validatePayload(payload, foldingOptionsSchemaDefinition));
export const loadCollection = createAsyncThunk('folding/loadCollection', async (collectionId, { getState, rejectWithValue, extra: { apiClient } }) => {
    const state = getState();
    const actualRequest = await buildInsightLoadCollectionRequest(state, collectionId);
    const fetched = await fetchFromAPI(apiClient, state, actualRequest, {
        origin: 'foldingCollection',
    });
    const response = fetched.response;
    if (isErrorResponse(response)) {
        return rejectWithValue(response.error);
    }
    return {
        collectionId,
        results: response.success.results,
        searchUid: response.success.searchUid,
        rootResult: state.folding.collections[collectionId]
            .result,
    };
});
