// @coveo/headless/dist/esm/features/insight-search/insight-query-suggest-request.js
import HistoryStore from '../../api/analytics/coveo.analytics/history-store.js';
import { getOrganizationEndpoint } from '../../api/platform-client.js';
import { fromAnalyticsStateToAnalyticsParams } from '../configuration/legacy-analytics-params.js';
export const buildInsightQuerySuggestRequest = async (id, s) => {
    return {
        accessToken: s.configuration.accessToken,
        organizationId: s.configuration.organizationId,
        url: getOrganizationEndpoint(s.configuration.organizationId, s.configuration.environment),
        count: s.querySuggest[id].count,
        insightId: s.insightConfiguration.insightId,
        q: s.querySet?.[id],
        timezone: s.configuration.search.timezone,
        actionsHistory: s.configuration.analytics.enabled
            ? HistoryStore.getInstance().getHistory()
            : [],
        ...(s.insightCaseContext?.caseContext && {
            context: s.insightCaseContext.caseContext,
        }),
        ...(s.configuration.analytics.enabled && {
            ...(s.configuration.analytics.enabled &&
                (await fromAnalyticsStateToAnalyticsParams(s.configuration.analytics))),
        }),
    };
};
