// @coveo/headless/dist/esm/features/insight-search/insight-analytics-actions.js
import { requiredNonEmptyString, validatePayload, } from '../../utils/validate-payload.js';
import { makeInsightAnalyticsActionFactory, } from '../analytics/analytics-utils.js';
import { SearchPageEvents } from '../analytics/search-action-cause.js';
import { getCaseContextAnalyticsMetadata } from '../case-context/case-context-state.js';
export const logExpandToFullUI = (fullSearchComponentName, triggeredBy) => makeInsightAnalyticsActionFactory(SearchPageEvents.expandToFullUI)({
    prefix: 'analytics/expandToFullUI',
    __legacy__getBuilder: (client, state) => {
        const metadata = getCaseContextAnalyticsMetadata(state.insightCaseContext);
        const meta = {
            caseId: metadata.caseId,
            caseNumber: metadata.caseNumber,
            fullSearchComponentName,
            triggeredBy,
            caseContext: metadata.caseContext,
        };
        return client.logExpandToFullUI(meta);
    },
    analyticsType: 'InsightPanel.ExpandToFullUI',
    analyticsPayloadBuilder: (state) => {
        const metadata = getCaseContextAnalyticsMetadata(state.insightCaseContext);
        return {
            context: {
                targetId: metadata.caseId,
                targetType: 'Case',
                caseNumber: metadata.caseNumber,
            },
        };
    },
});
export const logInsightCreateArticle = (createArticleMetadata) => makeInsightAnalyticsActionFactory(SearchPageEvents.createArticle)({
    prefix: 'analytics/insight/createArticle',
    __legacy__getBuilder: (client, state) => {
        validatePayload(createArticleMetadata, {
            articleType: requiredNonEmptyString,
        });
        return client.logCreateArticle(createArticleMetadata, getCaseContextAnalyticsMetadata(state.insightCaseContext));
    },
    analyticsType: 'InsightPanel.CreateArticle',
    analyticsPayloadBuilder: (state) => {
        const metadata = getCaseContextAnalyticsMetadata(state.insightCaseContext);
        return {
            articleType: createArticleMetadata.articleType,
            context: {
                targetId: metadata.caseId,
                targetType: 'Case',
                caseNumber: metadata.caseNumber,
            },
        };
    },
});
export const logOpenUserActions = () => makeInsightAnalyticsActionFactory('openUserActions')({
    prefix: 'analytics/insight/openUserActions',
    __legacy__getBuilder: (client, state) => {
        return client.logOpenUserActions(getCaseContextAnalyticsMetadata(state.insightCaseContext));
    },
});
