// @coveo/headless/dist/esm/controllers/insight/user-actions/headless-user-actions.js
import { configuration } from '../../../app/common-reducers.js';
import { logOpenUserActions } from '../../../features/insight-search/insight-analytics-actions.js';
import { fetchUserActions, registerUserActions, } from '../../../features/insight-user-actions/insight-user-actions-actions.js';
import { insightUserActionsReducer } from '../../../features/insight-user-actions/insight-user-actions-slice.js';
import { loadReducerError } from '../../../utils/errors.js';
import { buildController, } from '../../controller/headless-controller.js';
/**
 *
 * @param engine - The headless engine.
 * @param props - The configurable `UserActions` properties.
 * @returns A `UserActions` controller instance.
 *
 * @group Controllers
 * @category UserActions
 */
export function buildUserActions(engine, props) {
    if (!loadUserActionsReducers(engine)) {
        throw loadReducerError;
    }
    const { dispatch } = engine;
    const getState = () => engine.state.insightUserActions;
    const controller = buildController(engine);
    const { ticketCreationDate, excludedCustomActions } = props.options;
    dispatch(registerUserActions({ ticketCreationDate, excludedCustomActions }));
    return {
        ...controller,
        get state() {
            return getState();
        },
        fetchUserActions(userId) {
            dispatch(fetchUserActions(userId));
        },
        logOpenUserActions() {
            dispatch(logOpenUserActions());
        },
    };
}
function loadUserActionsReducers(engine) {
    engine.addReducers({
        configuration,
        insightUserActions: insightUserActionsReducer,
    });
    return true;
}
