// @coveo/headless/dist/esm/api/service/case-assist/case-assist-api-client.js
import { PlatformClient } from '../../platform-client.js';
import { getHtml, } from '../../search/html/html-api-client.js';
import { buildAPIResponseFromErrorOrThrow } from '../../search/search-api-error-response.js';
import { buildGetCaseClassificationsRequest, } from './get-case-classifications/get-case-classifications-request.js';
import { buildGetDocumentSuggestionsRequest, } from './get-document-suggestions/get-document-suggestions-request.js';
/**
 * The client to use to interface with the Case Assist API.
 */
export class CaseAssistAPIClient {
    options;
    constructor(options) {
        this.options = options;
    }
    /**
     * Retrieves the case classifications from the API.
     *
     * See https://platform.cloud.coveo.com/docs?urls.primaryName=Customer%20Service#/Suggestions/postClassify
     *
     * @param req - The request parameters.
     * @returns The case classifications grouped by fields for the given case information.
     */
    async getCaseClassifications(req) {
        const response = await PlatformClient.call({
            ...buildGetCaseClassificationsRequest(req),
            ...this.options,
        });
        if (response instanceof Error) {
            return buildAPIResponseFromErrorOrThrow(response);
        }
        const body = await response.json();
        return response.ok
            ? { success: body }
            : { error: body };
    }
    /**
     * Retrieves the document suggestions from the API.
     *
     * See https://platform.cloud.coveo.com/docs?urls.primaryName=Customer%20Service#/Suggestions/getSuggestDocument
     *
     * @param req - The request parameters.
     * @returns The document suggestions for the given case information and context.
     */
    async getDocumentSuggestions(req) {
        const response = await PlatformClient.call({
            ...buildGetDocumentSuggestionsRequest(req),
            ...this.options,
        });
        if (response instanceof Error) {
            return buildAPIResponseFromErrorOrThrow(response);
        }
        const body = await response.json();
        return response.ok
            ? {
                success: body,
            }
            : { error: body };
    }
    async html(req) {
        return getHtml(req, this.options);
    }
}
