// @coveo/headless/dist/esm/ssr-next/search/utils/controller-wiring.js
import { RecordValue, Schema } from '@coveo/bueno';
import { nonRequiredEmptyAllowedString } from '../../../utils/validate-payload.js';
const searchDefinition = {
    searchParams: new RecordValue({
        options: { required: false },
        values: { q: nonRequiredEmptyAllowedString },
    }),
};
export const searchDefinitionSchema = new Schema(searchDefinition);
/**
 * Validates the build configuration for search.
 */
function validateBuildConfig(buildConfig) {
    searchDefinitionSchema.validate(buildConfig);
}
/**
 * Controller wiring class that handles the complete wiring process for search.
 * Transforms simple user configuration into complex internal controller structures.
 */
class ControllerWirer {
    buildConfig;
    controllerDefinitions;
    controllerProps;
    constructor(buildConfig, controllerDefinitions, controllerProps) {
        this.buildConfig = buildConfig;
        this.controllerDefinitions = controllerDefinitions;
        this.controllerProps = controllerProps;
    }
    wireParameterManager() {
        if (!this.controllerDefinitions?.parameterManager)
            return;
        const { searchParams } = this.buildConfig;
        this.controllerProps.parameterManager = {
            initialState: {
                parameters: searchParams || {},
            },
        };
    }
    /**
     * Wires all controllers based on controller definitions.
     * Handles the complete wiring process in a sequential manner.
     */
    wire() {
        this.wireParameterManager();
    }
}
/**
 * Converts simple user configuration (searchParams) into the
 * nested structure required by the internal controller system. Automatically generates
 * appropriate controllers (parameterManager) transparently.
 *
 * @param controllerDefinitions - Map of controller definitions to be initialized
 * @param buildConfig - Simple configuration object with simplified controller properties
 * @returns Formatted and validated controller properties map defined for the controller in the definition
 */
export function wireControllerParams(controllerDefinitions, buildConfig) {
    validateBuildConfig(buildConfig);
    const controllerProps = buildConfig.controllers ?? {};
    new ControllerWirer(buildConfig, controllerDefinitions, controllerProps).wire();
    return controllerProps;
}
