// @coveo/headless/dist/esm/ssr-next/search/utils/engine-wiring.js
import { buildLogger } from '../../../app/logger.js';
import { augmentPreprocessRequestWithForwardedFor } from '../../common/augment-preprocess-request.js';
const ensureNavigatorContext = (buildConfig, loggerOptions) => {
    if (!buildConfig.navigatorContext) {
        const logger = buildLogger(loggerOptions);
        logger.error('No navigatorContext was provided. This may impact analytics accuracy, personalization, and session tracking. Refer to the Coveo documentation on server-side navigation context for implementation guidance.');
    }
};
export function augmentSearchEngineOptions(engineOptions, buildConfig) {
    ensureNavigatorContext(buildConfig, engineOptions.loggerOptions);
    return {
        ...engineOptions,
        navigatorContextProvider: () => buildConfig.navigatorContext,
        configuration: {
            ...engineOptions.configuration,
            preprocessRequest: augmentPreprocessRequestWithForwardedFor({
                preprocessRequest: engineOptions.configuration.preprocessRequest,
                navigatorContext: buildConfig.navigatorContext,
                loggerOptions: engineOptions.loggerOptions,
            }),
        },
    };
}
