// @coveo/headless/dist/esm/app/recommendation-engine/recommendation-engine.js
import { isNullOrUndefined } from '@coveo/bueno';
import { NoopPreprocessRequest } from '../../api/preprocess-request.js';
import { SearchAPIClient } from '../../api/search/search-api-client.js';
import { NoopPostprocessFacetSearchResponseMiddleware, NoopPostprocessQuerySuggestResponseMiddleware, NoopPostprocessSearchResponseMiddleware, } from '../../api/search/search-api-client-middleware.js';
import { updateSearchConfiguration } from '../../features/configuration/configuration-actions.js';
import { debugReducer as debug } from '../../features/debug/debug-slice.js';
import { setPipeline } from '../../features/pipeline/pipeline-actions.js';
import { pipelineReducer as pipeline } from '../../features/pipeline/pipeline-slice.js';
import { recommendationReducer as recommendation } from '../../features/recommendation/recommendation-slice.js';
import { setSearchHub } from '../../features/search-hub/search-hub-actions.js';
import { searchHubReducer as searchHub } from '../../features/search-hub/search-hub-slice.js';
import { buildEngine, } from '../engine.js';
import { buildLogger } from '../logger.js';
import { buildThunkExtraArguments } from '../thunk-extra-arguments.js';
import { getSampleRecommendationEngineConfiguration, recommendationEngineConfigurationSchema, } from './recommendation-engine-configuration.js';
export { getSampleRecommendationEngineConfiguration };
const recommendationEngineReducers = {
    debug,
    pipeline,
    searchHub,
    recommendation,
};
/**
 * Creates a recommendation engine instance.
 *
 * @param options - The recommendation engine options.
 * @returns A recommendation engine instance.
 *
 * @group Engine
 */
export function buildRecommendationEngine(options) {
    const logger = buildLogger(options.loggerOptions);
    validateConfiguration(options.configuration, logger);
    const searchAPIClient = createSearchAPIClient(options.configuration, logger);
    const thunkArguments = {
        ...buildThunkExtraArguments(options.configuration, logger),
        apiClient: searchAPIClient,
    };
    const augmentedOptions = {
        ...options,
        reducers: recommendationEngineReducers,
    };
    const engine = buildEngine(augmentedOptions, thunkArguments);
    const { pipeline, searchHub, timezone, locale, proxyBaseUrl } = options.configuration;
    engine.dispatch(updateSearchConfiguration({ timezone, locale, proxyBaseUrl }));
    if (!isNullOrUndefined(pipeline)) {
        engine.dispatch(setPipeline(pipeline));
    }
    if (!isNullOrUndefined(searchHub)) {
        engine.dispatch(setSearchHub(searchHub));
    }
    return {
        ...engine,
        get state() {
            return engine.state;
        },
    };
}
function validateConfiguration(configuration, logger) {
    try {
        recommendationEngineConfigurationSchema.validate(configuration);
    }
    catch (error) {
        logger.error(error, 'Recommendation engine configuration error');
        throw error;
    }
}
function createSearchAPIClient(configuration, logger) {
    return new SearchAPIClient({
        logger,
        preprocessRequest: configuration.preprocessRequest || NoopPreprocessRequest,
        postprocessSearchResponseMiddleware: configuration.preprocessSearchResponseMiddleware ||
            NoopPostprocessSearchResponseMiddleware,
        postprocessFacetSearchResponseMiddleware: NoopPostprocessFacetSearchResponseMiddleware,
        postprocessQuerySuggestResponseMiddleware: NoopPostprocessQuerySuggestResponseMiddleware,
    });
}
