// @coveo/headless/dist/esm/features/question-answering/question-answering-analytics-actions.js
import { validatePayload } from '../../utils/validate-payload.js';
import { documentIdentifier, makeAnalyticsAction, partialDocumentInformation, } from '../analytics/analytics-utils.js';
import { inlineLinkPayloadDefinition, uniqueIdentifierPayloadDefinition, validateQuestionAnsweringActionCreatorPayload, } from './question-answering-document-id.js';
import { answerSourceSelector, relatedQuestionSelector, } from './question-answering-selectors.js';
export const logExpandSmartSnippet = () => makeAnalyticsAction({
    prefix: 'analytics/smartSnippet/expand',
    __legacy__getBuilder: (client) => {
        return client.makeExpandSmartSnippet();
    },
    analyticsType: 'SmartSnippets.AnswerAction',
    analyticsPayloadBuilder: (state) => {
        const result = answerSourceSelector(state);
        const identifier = documentIdentifier(result);
        const searchUid = state.search?.response.searchUid;
        if (searchUid) {
            return {
                action: 'expand',
                snippetType: 'SmartSnippet',
                responseId: searchUid,
                itemMetadata: {
                    uniqueFieldName: identifier.contentIDKey,
                    uniqueFieldValue: identifier.contentIDValue,
                    title: result.title,
                    url: result.clickUri,
                },
            };
        }
    },
});
export const logCollapseSmartSnippet = () => makeAnalyticsAction({
    prefix: 'analytics/smartSnippet/collapse',
    __legacy__getBuilder: (client) => {
        return client.makeCollapseSmartSnippet();
    },
    analyticsType: 'SmartSnippets.AnswerAction',
    analyticsPayloadBuilder: (state) => {
        const result = answerSourceSelector(state);
        const identifier = documentIdentifier(result);
        const searchUid = state.search?.response.searchUid;
        if (searchUid) {
            return {
                action: 'collapse',
                snippetType: 'SmartSnippet',
                responseId: searchUid,
                itemMetadata: {
                    uniqueFieldName: identifier.contentIDKey,
                    uniqueFieldValue: identifier.contentIDValue,
                    title: result.title,
                    url: result.clickUri,
                },
            };
        }
    },
});
export const logLikeSmartSnippet = () => makeAnalyticsAction({
    prefix: 'analytics/smartSnippet/like',
    __legacy__getBuilder: (client) => {
        return client.makeLikeSmartSnippet();
    },
    analyticsType: 'SmartSnippets.AnswerAction',
    analyticsPayloadBuilder: (state) => {
        const result = answerSourceSelector(state);
        const identifier = documentIdentifier(result);
        const searchUid = state.search?.response.searchUid;
        if (searchUid) {
            return {
                action: 'like',
                snippetType: 'SmartSnippet',
                responseId: searchUid,
                itemMetadata: {
                    uniqueFieldName: identifier.contentIDKey,
                    uniqueFieldValue: identifier.contentIDValue,
                    title: result.title,
                    url: result.clickUri,
                },
            };
        }
    },
});
export const logDislikeSmartSnippet = () => makeAnalyticsAction({
    prefix: 'analytics/smartSnippet/dislike',
    __legacy__getBuilder: (client) => {
        return client.makeDislikeSmartSnippet();
    },
    analyticsType: 'SmartSnippets.AnswerAction',
    analyticsPayloadBuilder: (state) => {
        const result = answerSourceSelector(state);
        const identifier = documentIdentifier(result);
        const searchUid = state.search?.response.searchUid;
        if (searchUid) {
            return {
                action: 'dislike',
                snippetType: 'SmartSnippet',
                responseId: searchUid,
                itemMetadata: {
                    uniqueFieldName: identifier.contentIDKey,
                    uniqueFieldValue: identifier.contentIDValue,
                    title: result.title,
                    url: result.clickUri,
                },
            };
        }
    },
});
/**
 * @returns A dispatchable action.
 */
export const logOpenSmartSnippetSource = () => makeAnalyticsAction({
    prefix: 'analytics/smartSnippet/source/open',
    __legacy__getBuilder: (client, state) => {
        const result = answerSourceSelector(state);
        return client.makeOpenSmartSnippetSource(partialDocumentInformation(result, state), documentIdentifier(result));
    },
    analyticsType: 'SmartSnippets.SourceClick',
    analyticsPayloadBuilder: (state) => {
        const result = answerSourceSelector(state);
        const identifier = documentIdentifier(result);
        const searchUid = state.search?.response.searchUid;
        if (searchUid) {
            return {
                snippetType: 'SmartSnippet',
                responseId: searchUid,
                itemMetadata: {
                    uniqueFieldName: identifier.contentIDKey,
                    uniqueFieldValue: identifier.contentIDValue,
                    title: result.title,
                    url: result.clickUri,
                },
            };
        }
    },
});
export const logOpenSmartSnippetInlineLink = (payload) => makeAnalyticsAction({
    prefix: 'analytics/smartSnippet/source/open',
    __legacy__getBuilder: (client, state) => {
        validatePayload(payload, inlineLinkPayloadDefinition());
        const result = answerSourceSelector(state);
        if (!result) {
            return null;
        }
        return client.makeOpenSmartSnippetInlineLink(partialDocumentInformation(result, state), {
            ...documentIdentifier(result),
            ...payload,
        });
    },
    analyticsType: 'SmartSnippets.SourceClick',
    analyticsPayloadBuilder: (state) => {
        const result = answerSourceSelector(state);
        const identifier = documentIdentifier(result);
        const searchUid = state.search?.response.searchUid;
        if (searchUid) {
            return {
                snippetType: 'SmartSnippet',
                responseId: searchUid,
                itemMetadata: {
                    uniqueFieldName: identifier.contentIDKey,
                    uniqueFieldValue: identifier.contentIDValue,
                    title: result.title,
                    url: result.clickUri,
                },
            };
        }
    },
});
//TODO: SFINT-5435
export const logOpenSmartSnippetFeedbackModal = () => makeAnalyticsAction('analytics/smartSnippet/feedbackModal/open', (client) => client.makeOpenSmartSnippetFeedbackModal());
//TODO: SFINT-5435
export const logCloseSmartSnippetFeedbackModal = () => makeAnalyticsAction('analytics/smartSnippet/feedbackModal/close', (client) => client.makeCloseSmartSnippetFeedbackModal());
export const smartSnippetFeedbackMap = {
    does_not_answer: 'doesNotAnswer',
    partially_answers: 'partiallyAnswers',
    was_not_a_question: 'wasNotAQuestion',
};
export const logSmartSnippetFeedback = (feedback) => makeAnalyticsAction({
    prefix: 'analytics/smartSnippet/sendFeedback',
    __legacy__getBuilder: (client) => {
        return client.makeSmartSnippetFeedbackReason(feedback);
    },
    analyticsType: 'SmartSnippets.SubmitFeedback',
    analyticsPayloadBuilder: (state) => {
        const result = answerSourceSelector(state);
        const identifier = documentIdentifier(result);
        const searchUid = state.search?.response.searchUid;
        if (searchUid) {
            return {
                responseId: searchUid,
                snippetType: 'SmartSnippet',
                itemMetadata: {
                    uniqueFieldName: identifier.contentIDKey,
                    uniqueFieldValue: identifier.contentIDValue,
                    title: result.title,
                    url: result.clickUri,
                },
                reason: smartSnippetFeedbackMap[feedback],
            };
        }
    },
});
export const logSmartSnippetDetailedFeedback = (details) => makeAnalyticsAction({
    prefix: 'analytics/smartSnippet/sendFeedback',
    __legacy__getBuilder: (client) => {
        return client.makeSmartSnippetFeedbackReason('other', details);
    },
    analyticsType: 'SmartSnippets.SubmitFeedback',
    analyticsPayloadBuilder: (state) => {
        const result = answerSourceSelector(state);
        const identifier = documentIdentifier(result);
        const searchUid = state.search?.response.searchUid;
        if (searchUid) {
            return {
                responseId: searchUid,
                snippetType: 'SmartSnippet',
                itemMetadata: {
                    uniqueFieldName: identifier.contentIDKey,
                    uniqueFieldValue: identifier.contentIDValue,
                    title: result.title,
                    url: result.clickUri,
                },
                reason: 'other',
                additionalNotes: details,
            };
        }
    },
});
export const logExpandSmartSnippetSuggestion = (payload) => makeAnalyticsAction({
    prefix: 'analytics/smartSnippetSuggestion/expand',
    __legacy__getBuilder: (client, state) => {
        validateQuestionAnsweringActionCreatorPayload(payload);
        const relatedQuestion = relatedQuestionSelector(state, payload.questionAnswerId);
        if (!relatedQuestion) {
            return null;
        }
        return client.makeExpandSmartSnippetSuggestion({
            question: relatedQuestion.question,
            answerSnippet: relatedQuestion.answerSnippet,
            documentId: relatedQuestion.documentId,
        });
    },
    analyticsType: 'SmartSnippets.AnswerAction',
    analyticsPayloadBuilder: (state) => {
        const relatedQuestion = relatedQuestionSelector(state, payload.questionAnswerId);
        const searchUid = state.search?.response.searchUid;
        if (searchUid && relatedQuestion) {
            const source = answerSourceSelector(state, relatedQuestion.documentId);
            return {
                action: 'expand',
                responseId: searchUid,
                snippetType: 'SmartSnippetSuggestion',
                itemMetadata: {
                    uniqueFieldName: relatedQuestion.documentId.contentIdKey,
                    uniqueFieldValue: relatedQuestion.documentId.contentIdValue,
                    title: source?.title,
                    url: source?.clickUri,
                },
            };
        }
    },
});
export const logCollapseSmartSnippetSuggestion = (payload) => makeAnalyticsAction({
    prefix: 'analytics/smartSnippetSuggestion/collapse',
    __legacy__getBuilder: (client, state) => {
        validateQuestionAnsweringActionCreatorPayload(payload);
        const relatedQuestion = relatedQuestionSelector(state, payload.questionAnswerId);
        if (!relatedQuestion) {
            return null;
        }
        return client.makeCollapseSmartSnippetSuggestion({
            question: relatedQuestion.question,
            answerSnippet: relatedQuestion.answerSnippet,
            documentId: relatedQuestion.documentId,
        });
    },
    analyticsType: 'SmartSnippets.AnswerAction',
    analyticsPayloadBuilder: (state) => {
        const relatedQuestion = relatedQuestionSelector(state, payload.questionAnswerId);
        const searchUid = state.search?.response.searchUid;
        if (searchUid && relatedQuestion) {
            const source = answerSourceSelector(state, relatedQuestion.documentId);
            return {
                action: 'collapse',
                responseId: searchUid,
                snippetType: 'SmartSnippetSuggestion',
                itemMetadata: {
                    uniqueFieldName: relatedQuestion.documentId.contentIdKey,
                    uniqueFieldValue: relatedQuestion.documentId.contentIdValue,
                    title: source?.title,
                    url: source?.clickUri,
                },
            };
        }
    },
});
export const logOpenSmartSnippetSuggestionSource = (payload) => makeAnalyticsAction({
    prefix: 'analytics/smartSnippetSuggestion/source/open',
    __legacy__getBuilder: (client, state) => {
        validatePayload(payload, uniqueIdentifierPayloadDefinition());
        const relatedQuestion = relatedQuestionSelector(state, payload.questionAnswerId);
        if (!relatedQuestion) {
            return null;
        }
        const source = answerSourceSelector(state, relatedQuestion.documentId);
        if (!source) {
            return null;
        }
        return client.makeOpenSmartSnippetSuggestionSource(partialDocumentInformation(source, state), {
            question: relatedQuestion.question,
            answerSnippet: relatedQuestion.answerSnippet,
            documentId: relatedQuestion.documentId,
        });
    },
    analyticsType: 'SmartSnippets.SourceClick',
    analyticsPayloadBuilder: (state) => {
        const relatedQuestion = relatedQuestionSelector(state, payload.questionAnswerId);
        const searchUid = state.search?.response.searchUid;
        if (searchUid && relatedQuestion) {
            const source = answerSourceSelector(state, relatedQuestion.documentId);
            return {
                responseId: searchUid,
                snippetType: 'SmartSnippetSuggestion',
                itemMetadata: {
                    uniqueFieldName: relatedQuestion.documentId.contentIdKey,
                    uniqueFieldValue: relatedQuestion.documentId.contentIdValue,
                    title: source?.title,
                    url: source?.clickUri,
                },
            };
        }
    },
});
export const logOpenSmartSnippetSuggestionInlineLink = (identifier, link) => makeAnalyticsAction({
    prefix: 'analytics/smartSnippetSuggestion/source/open',
    __legacy__getBuilder: (client, state) => {
        validatePayload(identifier, uniqueIdentifierPayloadDefinition());
        validatePayload(link, inlineLinkPayloadDefinition());
        const relatedQuestion = relatedQuestionSelector(state, identifier.questionAnswerId);
        if (!relatedQuestion) {
            return null;
        }
        const source = answerSourceSelector(state, relatedQuestion.documentId);
        if (!source) {
            return null;
        }
        return client.makeOpenSmartSnippetSuggestionInlineLink(partialDocumentInformation(source, state), {
            question: relatedQuestion.question,
            answerSnippet: relatedQuestion.answerSnippet,
            documentId: relatedQuestion.documentId,
            linkText: link.linkText,
            linkURL: link.linkURL,
        });
    },
    analyticsType: 'SmartSnippets.SourceClick',
    analyticsPayloadBuilder: (state) => {
        const relatedQuestion = relatedQuestionSelector(state, identifier.questionAnswerId);
        const searchUid = state.search?.response.searchUid;
        if (searchUid && relatedQuestion) {
            const source = answerSourceSelector(state, relatedQuestion.documentId);
            return {
                responseId: searchUid,
                snippetType: 'SmartSnippetSuggestion',
                itemMetadata: {
                    uniqueFieldName: relatedQuestion.documentId.contentIdKey,
                    uniqueFieldValue: relatedQuestion.documentId.contentIdValue,
                    title: source?.title,
                    url: source?.clickUri,
                },
            };
        }
    },
});
export const smartSnippetAnalyticsClient = {
    logExpandSmartSnippet,
    logCollapseSmartSnippet,
    logLikeSmartSnippet,
    logDislikeSmartSnippet,
    logOpenSmartSnippetSource,
    logOpenSmartSnippetInlineLink,
    logOpenSmartSnippetFeedbackModal,
    logCloseSmartSnippetFeedbackModal,
    logSmartSnippetFeedback,
    logSmartSnippetDetailedFeedback,
    logExpandSmartSnippetSuggestion,
    logCollapseSmartSnippetSuggestion,
    logOpenSmartSnippetSuggestionSource,
};
