// @coveo/headless/dist/esm/controllers/insight/attached-results/headless-attached-results.js
import { isNullOrUndefined } from '@coveo/bueno';
import { configuration } from '../../../app/common-reducers.js';
import { attachResult, detachResult, } from '../../../features/attached-results/attached-results-actions.js';
import { logCaseAttach, logCaseDetach, } from '../../../features/attached-results/attached-results-analytics-actions.js';
import { attachedResultsReducer as attachedResults } from '../../../features/attached-results/attached-results-slice.js';
import { buildAttachedResultFromSearchResult } from '../../../features/attached-results/attached-results-utils.js';
import { loadReducerError } from '../../../utils/errors.js';
import { buildController, } from '../../controller/headless-controller.js';
/**
 * Creates an AttachedResults controller instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `AttachedResults` properties.
 * @returns A `AttachedResults` controller instance.
 *
 * @group Controllers
 * @category AttachedResults
 */
export function buildAttachedResults(engine, props) {
    if (!loadAttachedResultsReducers(engine)) {
        throw loadReducerError;
    }
    const { dispatch } = engine;
    const controller = buildController(engine);
    const { caseId } = props.options;
    const isResultAttached = (result) => {
        if (isNullOrUndefined(caseId)) {
            return false;
        }
        if (isNullOrUndefined(result.raw.permanentid) &&
            isNullOrUndefined(result.raw.urihash)) {
            return false;
        }
        return engine.state.attachedResults.results.some((attached) => {
            const caseIdMatches = caseId === attached.caseId;
            const permanentIdMatches = !isNullOrUndefined(attached.permanentId) &&
                attached.permanentId === result.raw.permanentid;
            const uriHashMatches = !isNullOrUndefined(attached.uriHash) &&
                attached.uriHash === result.raw.urihash;
            return caseIdMatches && (permanentIdMatches || uriHashMatches);
        });
    };
    const getAttachedResultsForRecord = () => {
        return engine.state.attachedResults.results.filter((attached) => attached.caseId === caseId);
    };
    return {
        ...controller,
        get state() {
            return {
                results: getAttachedResultsForRecord(),
                loading: engine.state.attachedResults.loading,
            };
        },
        isAttached(result) {
            return isResultAttached(result);
        },
        attach(result) {
            dispatch(attachResult(buildAttachedResultFromSearchResult(result, caseId)));
            dispatch(logCaseAttach(result));
        },
        detach(result) {
            dispatch(detachResult(buildAttachedResultFromSearchResult(result, caseId)));
            dispatch(logCaseDetach(result));
        },
    };
}
function loadAttachedResultsReducers(engine) {
    engine.addReducers({ configuration, attachedResults });
    return true;
}
