// @coveo/headless/dist/esm/api/analytics/recommendations-analytics.js
import { getQueryInitialState } from '../../features/query/query-state.js';
import { getRecommendationInitialState } from '../../features/recommendation/recommendation-state.js';
import { BaseAnalyticsProvider } from './base-analytics.js';
export class RecommendationAnalyticsProvider extends BaseAnalyticsProvider {
    getPipeline() {
        return (this.state.pipeline || this.state.recommendation?.pipeline || 'default');
    }
    getSearchEventRequestPayload() {
        return {
            queryText: getQueryInitialState().q,
            responseTime: this.responseTime,
            results: this.mapResultsToAnalyticsDocument(),
            numberOfResults: this.numberOfResults,
        };
    }
    getSearchUID() {
        const newState = this.getState();
        return (newState.recommendation?.searchUid ||
            getRecommendationInitialState().searchUid);
    }
    getSplitTestRunName() {
        return this.state.recommendation?.splitTestRun;
    }
    getSplitTestRunVersion() {
        const hasSplitTestRun = !!this.getSplitTestRunName();
        const effectivePipelineWithSplitTestRun = this.state.recommendation?.pipeline || this.state.pipeline || 'default';
        return hasSplitTestRun ? effectivePipelineWithSplitTestRun : undefined;
    }
    get responseTime() {
        return (this.state.recommendation?.duration ||
            getRecommendationInitialState().duration);
    }
    mapResultsToAnalyticsDocument() {
        return this.state.recommendation?.recommendations.map((r) => ({
            documentUri: r.uri,
            documentUriHash: r.raw.urihash,
        }));
    }
    get numberOfResults() {
        return (this.state.recommendation?.recommendations.length ||
            getRecommendationInitialState().recommendations.length);
    }
}
