// @coveo/headless/dist/esm/test/mock-commerce-facet-value.js
export function buildMockCommerceRegularFacetValue(config = {}) {
    return {
        value: '',
        state: 'idle',
        numberOfResults: 0,
        isAutoSelected: false,
        isSuggested: false,
        moreValuesAvailable: false,
        ...config,
    };
}
export function buildMockCommerceLocationFacetValue(config = {}) {
    return {
        value: '',
        state: 'idle',
        ...config,
    };
}
export function buildMockCommerceNumericFacetValue(config = {}) {
    return {
        end: 0,
        endInclusive: false,
        isAutoSelected: false,
        isSuggested: false,
        moreValuesAvailable: false,
        numberOfResults: 0,
        start: 0,
        state: 'idle',
        ...config,
    };
}
export function buildMockCommerceDateFacetValue(config = {}) {
    return {
        end: '',
        endInclusive: false,
        isAutoSelected: false,
        isSuggested: false,
        moreValuesAvailable: false,
        numberOfResults: 0,
        start: '',
        state: 'idle',
        ...config,
    };
}
export function buildMockCategoryFacetValue(config = {}) {
    return {
        children: [],
        isAutoSelected: false,
        isLeafValue: false,
        isSuggested: false,
        moreValuesAvailable: false,
        numberOfResults: 0,
        path: [],
        state: 'idle',
        value: '',
        ...config,
    };
}
