// @coveo/headless/dist/esm/test/mock-commerce-facet-response.js
function getMockBaseCommerceFacetResponse() {
    return {
        displayName: '',
        facetId: '',
        field: '',
        fromAutoSelect: false,
        isFieldExpanded: false,
        moreValuesAvailable: false,
        numberOfValues: 0,
    };
}
export function buildMockCommerceRegularFacetResponse(config = {}) {
    return {
        ...getMockBaseCommerceFacetResponse(),
        type: 'regular',
        values: [],
        ...config,
    };
}
export function buildMockCommerceLocationFacetResponse(config = {}) {
    return {
        ...getMockBaseCommerceFacetResponse(),
        type: 'location',
        values: [],
        ...config,
    };
}
export function buildMockCommerceNumericFacetResponse(config = {}) {
    return {
        ...getMockBaseCommerceFacetResponse(),
        type: 'numericalRange',
        values: [],
        interval: 'equiprobable',
        ...config,
    };
}
export function buildMockCommerceDateFacetResponse(config = {}) {
    return {
        ...getMockBaseCommerceFacetResponse(),
        type: 'dateRange',
        values: [],
        ...config,
    };
}
export function buildMockCategoryFacetResponse(config = {}) {
    return {
        ...getMockBaseCommerceFacetResponse(),
        type: 'hierarchical',
        values: [],
        delimitingCharacter: '',
        ...config,
    };
}
