// @coveo/headless/dist/esm/controllers/document-suggestion-list/headless-document-suggestion-list.js
import { configuration } from '../../app/common-reducers.js';
import { caseAssistConfigurationReducer as caseAssistConfiguration } from '../../features/case-assist-configuration/case-assist-configuration-slice.js';
import { caseFieldReducer as caseField } from '../../features/case-field/case-field-slice.js';
import { caseInputReducer as caseInput } from '../../features/case-input/case-input-slice.js';
import { fetchDocumentSuggestions } from '../../features/document-suggestion/document-suggestion-actions.js';
import { documentSuggestionReducer as documentSuggestion } from '../../features/document-suggestion/document-suggestion-slice.js';
import { loadReducerError } from '../../utils/errors.js';
import { buildController, } from '../controller/headless-controller.js';
/**
 * Creates a `Document Suggestion List` controller instance.
 *
 * @param engine - The headless engine.
 * @returns A `DocumentSuggestionList` controller instance.
 *
 * @group Controllers
 * @category DocumentSuggestionList
 */
export function buildDocumentSuggestionList(engine) {
    if (!loadDocumentSuggestionListReducers(engine)) {
        throw loadReducerError;
    }
    const controller = buildController(engine);
    const { dispatch } = engine;
    const getState = () => engine.state;
    return {
        ...controller,
        fetch() {
            dispatch(fetchDocumentSuggestions());
        },
        get state() {
            const state = getState().documentSuggestion;
            return {
                loading: state.status.loading,
                error: state.status.error,
                documents: state.documents,
            };
        },
    };
}
function loadDocumentSuggestionListReducers(engine) {
    engine.addReducers({
        configuration,
        caseAssistConfiguration,
        caseInput,
        caseField,
        documentSuggestion,
    });
    return true;
}
