// @coveo/headless/dist/esm/controllers/facets/range-facet/date-facet/headless-date-facet.js
import { facetClearAll, logFacetClearAll, logFacetUpdateSort, } from '../../../../features/facets/facet-set/facet-set-analytics-actions.js';
import { getAnalyticsActionForToggleFacetSelect, getLegacyAnalyticsActionForToggleRangeFacetExclude, getLegacyAnalyticsActionForToggleRangeFacetSelect, } from '../../../../features/facets/range-facets/generic/range-facet-utils.js';
import { executeSearch } from '../../../../features/search/search-actions.js';
import { buildCoreDateFacet, buildDateRange, } from '../../../core/facets/range-facet/date-facet/headless-core-date-facet.js';
export { buildDateRange };
/**
 * Creates a `DateFacet` controller instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `DateFacet` controller properties.
 * @returns A `DateFacet` controller instance.
 *
 * @group Controllers
 * @category DateFacet
 */
export function buildDateFacet(engine, props) {
    const coreController = buildCoreDateFacet(engine, props);
    const dispatch = engine.dispatch;
    const getFacetId = () => coreController.state.facetId;
    return {
        ...coreController,
        deselectAll() {
            coreController.deselectAll();
            dispatch(executeSearch({
                legacy: logFacetClearAll(getFacetId()),
                next: facetClearAll(),
            }));
        },
        sortBy(criterion) {
            coreController.sortBy(criterion);
            dispatch(executeSearch({
                legacy: logFacetUpdateSort({ facetId: getFacetId(), criterion }),
            }));
        },
        toggleSelect: (selection) => {
            coreController.toggleSelect(selection);
            dispatch(executeSearch({
                legacy: getLegacyAnalyticsActionForToggleRangeFacetSelect(getFacetId(), selection),
                next: getAnalyticsActionForToggleFacetSelect(selection),
            }));
        },
        toggleExclude: (selection) => {
            coreController.toggleExclude(selection);
            dispatch(executeSearch({
                legacy: getLegacyAnalyticsActionForToggleRangeFacetExclude(getFacetId(), selection),
            }));
        },
        get state() {
            return coreController.state;
        },
    };
}
