// @coveo/headless/dist/esm/controllers/facets/range-facet/date-facet/headless-date-filter.js
import { configuration } from '../../../../app/common-reducers.js';
import { facetClearAll, facetSelect, logFacetClearAll, logFacetSelect, } from '../../../../features/facets/facet-set/facet-set-analytics-actions.js';
import { dateFacetSetReducer as dateFacetSet } from '../../../../features/facets/range-facets/date-facet-set/date-facet-set-slice.js';
import { executeSearch } from '../../../../features/search/search-actions.js';
import { searchReducer as search } from '../../../../features/search/search-slice.js';
import { loadReducerError } from '../../../../utils/errors.js';
import { buildCoreDateFilter, } from '../../../core/facets/range-facet/date-facet/headless-core-date-filter.js';
/**
 * Creates a `DateFilter` controller instance.
 * @param engine - The headless engine.
 * @param props - The configurable `DateFilter` controller properties.
 * @returns A `DateFilter` controller instance.
 *
 * @group Controllers
 * @category DateFilter
 */
export function buildDateFilter(engine, props) {
    if (!loadDateFilterReducer(engine)) {
        throw loadReducerError;
    }
    const coreController = buildCoreDateFilter(engine, props);
    const { dispatch } = engine;
    const getFacetId = () => coreController.state.facetId;
    return {
        ...coreController,
        clear: () => {
            coreController.clear();
            dispatch(executeSearch({
                legacy: logFacetClearAll(getFacetId()),
                next: facetClearAll(),
            }));
        },
        setRange: (range) => {
            const success = coreController.setRange(range);
            if (success) {
                dispatch(executeSearch({
                    legacy: logFacetSelect({
                        facetId: getFacetId(),
                        facetValue: `${range.start}..${range.end}`,
                    }),
                    next: facetSelect(),
                }));
            }
            return success;
        },
        get state() {
            return {
                ...coreController.state,
            };
        },
    };
}
function loadDateFilterReducer(engine) {
    engine.addReducers({ dateFacetSet, configuration, search });
    return true;
}
