// @coveo/headless/dist/esm/controllers/insight/smart-snippet-questions-list/headless-insight-smart-snippet-questions-list.js
import { insightSmartSnippetAnalyticsClient } from '../../../features/question-answering/question-answering-insight-analytics-actions.js';
import { buildCoreSmartSnippetQuestionsList, } from '../../core/smart-snippet-questions-list/headless-core-smart-snippet-questions-list.js';
import { buildSmartSnippetInteractiveInlineLinks } from '../smart-snippet/headless-insight-smart-snippet-interactive-inline-links.js';
import { buildInsightSmartSnippetInteractiveQuestions } from './headless-insight-smart-snippet-interactive-questions.js';
/**
 * Creates an insight `SmartSnippetQuestionsList` controller instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `SmartSnippetQuestionsList` properties.
 * @returns A `SmartSnippetQuestionsList` controller instance.
 *
 * @group Controllers
 * @category SmartSnippetQuestionsList
 * */
export function buildSmartSnippetQuestionsList(engine, props) {
    const smartSnippetQuestionList = buildCoreSmartSnippetQuestionsList(engine, insightSmartSnippetAnalyticsClient);
    const interactiveInlineLinks = buildSmartSnippetInteractiveInlineLinks(engine, {
        options: { selectionDelay: props?.options?.selectionDelay },
    });
    const interactiveQuestions = buildInsightSmartSnippetInteractiveQuestions(engine, {
        options: { selectionDelay: props?.options?.selectionDelay },
    });
    return {
        ...smartSnippetQuestionList,
        get state() {
            return smartSnippetQuestionList.state;
        },
        selectSource(identifier) {
            interactiveQuestions.selectSource(identifier);
        },
        beginDelayedSelectSource(identifier) {
            interactiveQuestions.beginDelayedSelectSource(identifier);
        },
        cancelPendingSelectSource(identifier) {
            interactiveQuestions.cancelPendingSelectSource(identifier);
        },
        selectInlineLink(identifier, link) {
            interactiveInlineLinks.selectInlineLink(link, identifier);
        },
        beginDelayedSelectInlineLink(identifier, link) {
            interactiveInlineLinks.beginDelayedSelectInlineLink(link, identifier);
        },
        cancelPendingSelectInlineLink(identifier, link) {
            interactiveInlineLinks.cancelPendingSelectInlineLink(link, identifier);
        },
    };
}
