// @coveo/headless/dist/esm/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.js
import { facetClearAll } from '../../../../../features/facets/facet-set/facet-set-analytics-actions.js';
import { logFacetClearAll, logFacetUpdateSort, } from '../../../../../features/facets/facet-set/facet-set-insight-analytics-actions.js';
import { getInsightAnalyticsActionForToggleRangeFacetSelect } from '../../../../../features/facets/range-facets/generic/range-facet-insight-utils.js';
import { getAnalyticsActionForToggleFacetSelect } from '../../../../../features/facets/range-facets/generic/range-facet-utils.js';
import { executeSearch } from '../../../../../features/insight-search/insight-search-actions.js';
import { buildCoreDateFacet, buildDateRange, } from '../../../../core/facets/range-facet/date-facet/headless-core-date-facet.js';
export { buildDateRange };
/**
 * Creates an insight `DateFacet` controller instance.
 *
 * @param engine - The insight engine.
 * @param props - The configurable `DateFacet` controller properties.
 * @returns A `DateFacet` controller instance.
 *
 * @group Controllers
 * @category DateFacet
 */
export function buildDateFacet(engine, props) {
    const coreController = buildCoreDateFacet(engine, props);
    const dispatch = engine.dispatch;
    const getFacetId = () => coreController.state.facetId;
    return {
        ...coreController,
        deselectAll() {
            coreController.deselectAll();
            dispatch(executeSearch({
                legacy: logFacetClearAll(getFacetId()),
                next: facetClearAll(),
            }));
        },
        sortBy(sortCriterion) {
            coreController.sortBy(sortCriterion);
            dispatch(executeSearch({
                legacy: logFacetUpdateSort({ facetId: getFacetId(), sortCriterion }),
            }));
        },
        toggleSelect: (selection) => {
            coreController.toggleSelect(selection);
            dispatch(executeSearch({
                legacy: getInsightAnalyticsActionForToggleRangeFacetSelect(getFacetId(), selection),
                next: getAnalyticsActionForToggleFacetSelect(selection),
            }));
        },
        get state() {
            return coreController.state;
        },
    };
}
