// @coveo/headless/dist/esm/features/result-preview/result-preview-analytics-actions.js
import { documentIdentifier, makeAnalyticsAction, partialDocumentInformation, validateResultPayload, } from '../analytics/analytics-utils.js';
export const logDocumentQuickview = (result) => {
    return makeAnalyticsAction({
        prefix: 'analytics/resultPreview/open',
        __legacy__getBuilder: (client, state) => {
            validateResultPayload(result);
            const info = partialDocumentInformation(result, state);
            const id = documentIdentifier(result);
            return client.makeDocumentQuickview(info, id);
        },
        analyticsType: 'itemClick',
        analyticsPayloadBuilder: (state) => {
            const docInfo = partialDocumentInformation(result, state);
            const docId = documentIdentifier(result);
            return {
                responseId: result.searchUid ?? '',
                position: docInfo.documentPosition,
                itemMetadata: {
                    uniqueFieldName: docId.contentIDKey,
                    uniqueFieldValue: docId.contentIDValue,
                    title: docInfo.documentTitle,
                    author: docInfo.documentAuthor,
                    url: docInfo.documentUrl,
                },
            };
        },
    });
};
