// @coveo/headless/dist/esm/features/result-preview/result-preview-insight-analytics-actions.js
import { analyticsEventItemMetadata, documentIdentifier, makeInsightAnalyticsActionFactory, partialDocumentInformation, validateResultPayload, } from '../analytics/analytics-utils.js';
import { analyticsEventCaseContext } from '../analytics/insight-analytics-utils.js';
export const logDocumentQuickview = (result) => {
    return makeInsightAnalyticsActionFactory('')({
        prefix: 'analytics/resultPreview/open',
        __legacy__getBuilder: (client, state) => {
            validateResultPayload(result);
            const info = partialDocumentInformation(result, state);
            const id = documentIdentifier(result);
            client.logDocumentQuickview(info, id);
        },
        analyticsType: 'InsightPanel.ItemAction',
        analyticsPayloadBuilder: (state) => {
            const information = partialDocumentInformation(result, state);
            return {
                action: 'preview',
                itemMetadata: analyticsEventItemMetadata(result, state),
                position: information.documentPosition,
                responseId: result.searchUid || '',
                context: analyticsEventCaseContext(state),
            };
        },
    });
};
