// @coveo/headless/dist/esm/controllers/insight/search-box/headless-insight-search-box.js
import { executeSearch, fetchQuerySuggestions, } from '../../../features/insight-search/insight-search-actions.js';
import { searchboxSubmit } from '../../../features/query/query-analytics-actions.js';
import { logSearchboxSubmit } from '../../../features/query/query-insight-analytics-actions.js';
import { buildCoreSearchBox, } from '../../core/search-box/headless-core-search-box.js';
/**
 * Creates an insight `SearchBox` controller instance.
 *
 * @param engine - The insight engine.
 * @param props - The configurable `SearchBox` properties.
 * @returns A `SearchBox` controller instance.
 *
 * @group Controllers
 * @category SearchBox
 */
export function buildSearchBox(engine, props = {}) {
    const searchBox = buildCoreSearchBox(engine, {
        ...props,
        executeSearchActionCreator: executeSearch,
        fetchQuerySuggestionsActionCreator: fetchQuerySuggestions,
        isNextAnalyticsReady: true,
    });
    return {
        ...searchBox,
        submit() {
            searchBox.submit(logSearchboxSubmit(), searchboxSubmit());
        },
        get state() {
            return searchBox.state;
        },
    };
}
