// @coveo/headless/dist/esm/features/insight-interface/insight-interface-actions.js
import { createAsyncThunk } from '@reduxjs/toolkit';
import { getOrganizationEndpoint } from '../../api/platform-client.js';
import { isErrorResponse } from '../../api/search/search-api-client.js';
import { setSearchHub } from '../search-hub/search-hub-actions.js';
export const fetchInterface = createAsyncThunk('insight/interface/fetch', async (_, { getState, dispatch, rejectWithValue, extra: { apiClient } }) => {
    const state = getState();
    const fetched = await apiClient.getInterface(buildGetInsightInterfaceRequest(state));
    if (isErrorResponse(fetched)) {
        return rejectWithValue(fetched.error);
    }
    dispatch(setSearchHub(fetched.success.searchHub));
    return {
        response: fetched.success,
    };
});
const buildGetInsightInterfaceRequest = (state) => ({
    accessToken: state.configuration.accessToken,
    organizationId: state.configuration.organizationId,
    url: getOrganizationEndpoint(state.configuration.organizationId, state.configuration.environment),
    insightId: state.insightConfiguration.insightId,
});
