// @coveo/headless/dist/esm/features/insight-user-actions/insight-user-actions-actions.js
import { ArrayValue, StringValue } from '@coveo/bueno';
import { createAction, createAsyncThunk } from '@reduxjs/toolkit';
import { isErrorResponse } from '../../api/search/search-api-client.js';
import { nonEmptyString, validatePayload } from '../../utils/validate-payload.js';
import { buildFetchUserActionsRequest } from './insight-user-actions-request.js';
const registerUserActionsPayloadSchema = {
    ticketCreationDate: new StringValue({
        emptyAllowed: false,
        ISODate: true,
    }),
    excludedCustomActions: new ArrayValue({
        required: false,
        each: nonEmptyString,
    }),
};
export const registerUserActions = createAction('insight/userActions/registerUserActions', (payload) => validatePayload(payload, registerUserActionsPayloadSchema));
export const fetchUserActions = createAsyncThunk('insight/userActions/fetch', async (userId, { getState, rejectWithValue, extra: { apiClient } }) => {
    const state = getState();
    const fetched = await apiClient.userActions(await buildFetchUserActionsRequest(state, userId));
    if (isErrorResponse(fetched)) {
        return rejectWithValue(fetched.error);
    }
    return {
        response: fetched.success,
    };
});
