// @coveo/headless/dist/esm/controllers/case-input/headless-case-input.js
import { Schema } from '@coveo/bueno';
import { configuration } from '../../app/common-reducers.js';
import { logUpdateCaseField } from '../../features/case-assist/case-assist-analytics-actions.js';
import { caseAssistConfigurationReducer as caseAssistConfiguration } from '../../features/case-assist-configuration/case-assist-configuration-slice.js';
import { fetchCaseClassifications } from '../../features/case-field/case-field-actions.js';
import { caseFieldReducer as caseField } from '../../features/case-field/case-field-slice.js';
import { updateCaseInput } from '../../features/case-input/case-input-actions.js';
import { caseInputReducer as caseInput } from '../../features/case-input/case-input-slice.js';
import { fetchDocumentSuggestions } from '../../features/document-suggestion/document-suggestion-actions.js';
import { documentSuggestionReducer as documentSuggestion } from '../../features/document-suggestion/document-suggestion-slice.js';
import { loadReducerError } from '../../utils/errors.js';
import { requiredNonEmptyString, validateOptions, } from '../../utils/validate-payload.js';
import { buildController, } from '../controller/headless-controller.js';
function validateCaseInputOptions(engine, options) {
    const schema = new Schema({
        field: requiredNonEmptyString,
    });
    validateOptions(engine, schema, options, 'buildCaseInput');
}
/**
 * Creates a `Case Input` controller instance.
 *
 * @param engine - The headless engine.
 * @param props - The configurable `CaseInput` controller properties.
 * @returns A `CaseInput` controller instance.
 *
 * @group Controllers
 * @category CaseInput
 */
export function buildCaseInput(engine, props) {
    if (!loadCaseInputReducers(engine)) {
        throw loadReducerError;
    }
    const controller = buildController(engine);
    const { dispatch } = engine;
    const getState = () => engine.state;
    validateCaseInputOptions(engine, props.options);
    const fieldName = props.options.field;
    const isRegistered = getState().caseInput?.[fieldName];
    if (!isRegistered) {
        dispatch(updateCaseInput({
            fieldName: fieldName,
            fieldValue: '',
        }));
    }
    return {
        ...controller,
        update(value, updatesToFetch) {
            dispatch(updateCaseInput({
                fieldName: fieldName,
                fieldValue: value,
            }));
            dispatch(logUpdateCaseField(fieldName));
            updatesToFetch?.caseClassifications &&
                dispatch(fetchCaseClassifications());
            updatesToFetch?.documentSuggestions &&
                dispatch(fetchDocumentSuggestions());
        },
        get state() {
            return getState().caseInput[fieldName];
        },
    };
}
function loadCaseInputReducers(engine) {
    engine.addReducers({
        configuration,
        caseAssistConfiguration,
        caseInput,
        caseField,
        documentSuggestion,
    });
    return true;
}
